<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<style>body{background-color:white;}</style>
<link href="lib/htmltools-fill-0.5.9/fill.css" rel="stylesheet" />
<script src="lib/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="lib/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="lib/datatables-binding-0.34.0/datatables.js"></script>
<script src="lib/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="lib/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="lib/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="lib/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="lib/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet" />
<script src="lib/crosstalk-1.2.2/js/crosstalk.min.js"></script>
<link href="lib/highlight.js-6.2/rstudio.css" rel="stylesheet" />
<script src="lib/highlight.js-6.2/highlight.pack.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="lib/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<link href="lib/bootstrap-3.3.5/css/bootstrap-theme.min.css" rel="stylesheet" />
<script src="lib/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="lib/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="lib/bootstrap-3.3.5/shim/respond.min.js"></script>

</head>
<body>
<div class="container-fluid">
  <style type="text/css">table tr:hover td {
  font-weight:bold;text-decoration:none
}
table tr.covered td{
  background-color:rgba(95,151,68,0.3)
}

table tr:hover.covered .num{
  background-color:rgba(95,151,68,0.7)
}
table tr.missed td{
  background-color:rgba(185,73,71,0.3)
}
table tr:hover.missed .num{
  background-color:rgba(185,73,71,0.7)
}

table tr.missed:hover td{
  -webkit-box-shadow:0 -2px 0 0 #b94947 inset;
  -moz-box-shadow:0 -2px 0 0 #b94947 inset;
  box-shadow:0 -2px 0 0 #b94947 inset
}
table tr.covered:hover td{
  -webkit-box-shadow:0 -2px 0 0 #5f9744 inset;
  -moz-box-shadow:0 -2px 0 0 #5f9744 inset;
  box-shadow:0 -2px 0 0 #5f9744 inset
}

table tr.never td{
  background-color:transparent
}

table tbody {
  border-style: solid;
  border: 1px solid rgba(0,0,0,0.1)
}

table .num {
  border-right: 1px solid rgba(0,0,0,0.1)
}

td.coverage em {
  opacity: 0.5;
}

table td.coverage {
  border-right: 1px solid rgba(0,0,0,0.1);
  font-weight: bold;
  text-align: center;
}
table.table-condensed pre {
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 11px;
}
div#files td {
  padding: 0;
  padding-left: 5px;
}

div#files td.num {
  padding-right: 5px;
}

table.table-condensed {
  font-size: 11px;
}</style>
  <div class="col-md-8 col-md-offset-2">
    <h2>TSENAT coverage - 93.31%</h2>
    <div class="tabbable">
      <ul class="nav nav-tabs" data-tabsetid="covr">
        <li class="active">
          <a href="#tab-covr-1" data-toggle="tab" data-value="Files">Files</a>
        </li>
        <li>
          <a href="#tab-covr-2" data-toggle="tab" data-value="Source">Source</a>
        </li>
      </ul>
      <div class="tab-content" data-tabsetid="covr">
        <div class="tab-pane active" title="Files" data-value="Files" id="tab-covr-1">
          <div class="datatables html-widget html-fill-item" id="htmlwidget-159c74a742c9a259cf14" style="width:100%;height:500px;"></div>
          <script type="application/json" data-for="htmlwidget-159c74a742c9a259cf14">{"x":{"filter":"none","vertical":false,"fillContainer":false,"data":[["<a href=\"#\">R/calc_lm_helpers.R<\/a>","<a href=\"#\">R/map_functions.R<\/a>","<a href=\"#\">R/calculate_difference.R<\/a>","<a href=\"#\">R/filter_se.R<\/a>","<a href=\"#\">R/diversity_functions.R<\/a>","<a href=\"#\">R/generate_plots.R<\/a>","<a href=\"#\">R/plot_top_helpers.R<\/a>","<a href=\"#\">R/calculate_method.R<\/a>","<a href=\"#\">R/difference_functions.R<\/a>","<a href=\"#\">R/plot_helpers.R<\/a>","<a href=\"#\">R/diversity_helpers.R<\/a>","<a href=\"#\">R/difference_helpers.R<\/a>","<a href=\"#\">R/calculate_diversity.R<\/a>","<a href=\"#\">R/build_se.R<\/a>","<a href=\"#\">R/calculate_method_helpers.R<\/a>"],[568,388,304,111,81,1156,160,55,237,106,122,161,108,65,15],[247,194,118,47,34,527,113,27,82,67,75,113,48,24,8],[218,172,105,42,31,497,108,26,79,65,73,110,47,24,8],[29,22,13,5,3,30,5,1,3,2,2,3,1,0,0],["10","11","12","4","186","6","7","25","146","11","136","95","15","7","165"],["88.26%","88.66%","88.98%","89.36%","91.18%","94.31%","95.58%","96.30%","96.34%","97.01%","97.33%","97.35%","97.92%","100.00%","100.00%"]],"container":"<table class=\"row-border\">\n  <thead>\n    <tr>\n      <th>File<\/th>\n      <th>Lines<\/th>\n      <th>Relevant<\/th>\n      <th>Covered<\/th>\n      <th>Missed<\/th>\n      <th>Hits / Line<\/th>\n      <th>Coverage<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"searching":false,"dom":"t","paging":false,"columnDefs":[{"targets":6,"createdCell":"function(td, cellData, rowData, row, col) {\n  var percent = cellData.replace(\"%\", \"\");\n  if (percent > 90) {\n    var grad = \"linear-gradient(90deg, #edfde7 \" + cellData + \", white \" + cellData + \")\";\n  } else if (percent > 75) {\n    var grad = \"linear-gradient(90deg, #f9ffe5 \" + cellData + \", white \" + cellData + \")\";\n  } else {\n    var grad = \"linear-gradient(90deg, #fcece9 \" + cellData + \", white \" + cellData + \")\";\n  }\n  $(td).css(\"background\", grad);\n}\n"},{"className":"dt-right","targets":[1,2,3,4]},{"name":"File","targets":0},{"name":"Lines","targets":1},{"name":"Relevant","targets":2},{"name":"Covered","targets":3},{"name":"Missed","targets":4},{"name":"Hits / Line","targets":5},{"name":"Coverage","targets":6}],"order":[],"autoWidth":false,"orderClasses":false},"callback":"function(table) {\ntable.on('click.dt', 'a', function() {\n  files = $('div#files div');\n  files.not('div.hidden').addClass('hidden');\n  id = $(this).text();\n  files.filter('div[id=\\'' + id + '\\']').removeClass('hidden');\n  $('ul.nav a[data-value=Source]').text(id).tab('show');\n});\n}"},"evals":["options.columnDefs.0.createdCell","callback"],"jsHooks":[]}</script>
        </div>
        <div class="tab-pane" title="Source" data-value="Source" id="tab-covr-2">
          <div id="files">
            <div id="R/build_se.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Helper: build SummarizedExperiment from readcounts + tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Build a SummarizedExperiment from transcript readcounts and tx-&gt;gene map</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' This helper creates a `SummarizedExperiment` with an assay named</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `counts`, stores the `tx2gene` table and the raw `readcounts` in</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `metadata()`, and places the provided `genes` vector into `rowData(se)$genes`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param tx2gene_tsv Path to a tab-separated `tx2gene` file or a data.frame</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   with transcript-to-gene mapping. If a path is provided, it must contain</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   at least two columns (transcript, gene).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param readcounts Numeric matrix or data.frame of transcript-level counts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   (rows = transcripts, columns = samples).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param genes Character vector of gene IDs assigning each transcript to a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   gene; length must equal nrow(readcounts).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name for the assay to store readcounts (default: 'counts').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `SummarizedExperiment` with assay, `metadata()$tx2gene`,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `metadata()$readcounts` and `rowData(se)$genes` populated.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' tx2gene &lt;- data.frame(Transcript = c('tx1', 'tx2'), Gene = c('g1', 'g2'), stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' readcounts &lt;- matrix(c(10, 5, 2, 3), nrow = 2, dimnames = list(c('tx1', 'tx2'), c('s1', 's2')))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' genes &lt;- c('g1', 'g2')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- build_se(tx2gene, readcounts, genes)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' SummarizedExperiment::assay(se, 'counts')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">build_se &lt;- function(tx2gene_tsv, readcounts, genes, assay_name = "counts") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">26</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(tx2gene_tsv) &amp;&amp; length(tx2gene_tsv) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!file.exists(tx2gene_tsv)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("tx2gene file not found: ", tx2gene_tsv, call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tx2gene_df &lt;- utils::read.table(tx2gene_tsv, header = TRUE, sep = "\t", stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.data.frame(tx2gene_tsv)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tx2gene_df &lt;- tx2gene_tsv</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("'tx2gene_tsv' must be a path or a data.frame.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.data.frame(readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        readcounts &lt;- as.matrix(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">39</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.matrix(readcounts) || !is.numeric(readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("'readcounts' must be a numeric matrix or numeric data.frame.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">43</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(genes) != nrow(readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Length of 'genes' must equal nrow(readcounts).", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">48</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assays_list &lt;- S4Vectors::SimpleList()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assays_list[[assay_name]] &lt;- readcounts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    se &lt;- SummarizedExperiment::SummarizedExperiment(assays = assays_list)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    S4Vectors::metadata(se)$tx2gene &lt;- tx2gene_df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    S4Vectors::metadata(se)$readcounts &lt;- readcounts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # keep gene assignment in rowData for compatibility with other helpers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # ensure rownames align with transcript IDs if present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rn &lt;- rownames(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(rn)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        SummarizedExperiment::rowData(se) &lt;- S4Vectors::DataFrame(genes = genes)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">59</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        SummarizedExperiment::rowData(se) &lt;- S4Vectors::DataFrame(genes = genes,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            row.names = rn)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">62</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">65</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/generate_plots.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># # Helpers for generating diversity plots (density, violin, MA)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom utils head</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom grDevices png dev.off</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">if (getRversion() &gt;= "2.15.1") {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    utils::globalVariables(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        c(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "Gene",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "sample",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "sample_q",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "sample_type",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "fold",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "significant",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "value",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ".",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "group",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "tsallis",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "q",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "median",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "IQR",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "padj_num",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "padj_clean",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "xval",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "label_flag",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "genes",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "mean",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # transcript plotting globals</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "tx",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "expr",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">31</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "tx_cond",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "log2expr",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # generic plotting helpers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "x",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "y",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "padj"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">39</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">40</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs &lt;- function(pkgs) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">78<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        missing &lt;- pkgs[!vapply(pkgs, requireNamespace, logical(1), quietly = TRUE)]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">42</td>
                    <td class="coverage">78<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(missing)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                sprintf(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    "%s required",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">46</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    paste(missing, collapse = ", ")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">77<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        invisible(TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Small utility helpers to reduce repeated code paths when extracting</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># samples, readcounts and tx-&gt;gene mappings from a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># SummarizedExperiment. Keeping these as focused helpers improves</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># readability of the longer plotting functions below.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">58</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">infer_samples_from_se &lt;- function(se, samples = NULL, sample_type_col = "sample_type") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(samples)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(as.character(samples))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    cd &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    try(cd &lt;- SummarizedExperiment::colData(se), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(cd)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">65</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">66</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">67</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">68</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Common column names to try</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    candidates &lt;- c(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_type_col,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "condition",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "group",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "sample_group",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "sampleType",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "class",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "status",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "phenotype"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">78</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (nm in candidates) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">49<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (nm %in% colnames(cd)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.character(cd[[nm]]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">83</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">84</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">85</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Fallback: choose column with smallest &gt;1 unique values</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uniq_counts &lt;- vapply(cd, function(col) length(unique(na.omit(col))), integer(1))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    valid_cols &lt;- names(uniq_counts[uniq_counts &gt; 1])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(valid_cols) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        bin_cols &lt;- valid_cols[uniq_counts[valid_cols] == 2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pick &lt;- if (length(bin_cols) &gt; 0) bin_cols[1] else valid_cols[which.min(uniq_counts[valid_cols])]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(as.character(cd[[pick]]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">92</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">93</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">94</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">95</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">96</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">97</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">get_readcounts_from_se &lt;- function(se, readcounts_arg = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If user provided a readcounts object/path, accept it first</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(readcounts_arg)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.character(readcounts_arg) &amp;&amp; length(readcounts_arg) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">101</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!file.exists(readcounts_arg)) stop("readcounts file not found: ", readcounts_arg)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            rc_df &lt;- utils::read.delim(readcounts_arg, header = TRUE, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.null(colnames(rc_df)) &amp;&amp; ncol(rc_df) &gt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                counts &lt;- as.matrix(rc_df[, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                rownames(counts) &lt;- rc_df[[1]]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                counts &lt;- as.matrix(rc_df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">108</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">109</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(counts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">110</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if (is.matrix(readcounts_arg) || is.data.frame(readcounts_arg)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.matrix(readcounts_arg))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">112</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">113</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("`readcounts` must be a matrix/data.frame or path to a file")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">114</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">115</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">116</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">117</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    md &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">118</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    try(md &lt;- S4Vectors::metadata(se), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">119</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(md) &amp;&amp; !is.null(md$readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">120</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(as.matrix(md$readcounts))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">121</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">122</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">123</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assay_names &lt;- SummarizedExperiment::assayNames(se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">124</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    preferred_assays &lt;- c("readcounts", "counts", "tx_counts", "counts_tx")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    chosen &lt;- intersect(preferred_assays, assay_names)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">126</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(chosen) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">127</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(as.matrix(SummarizedExperiment::assay(se, chosen[1])))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">128</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">129</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">130</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # fallback to first assay with a warning</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">131</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    warning(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">132</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "Using first assay from SummarizedExperiment to compute",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">133</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        " expression-based fold changes; ensure it contains",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">134</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        " transcript-level readcounts or provide metadata$readcounts"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">135</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">136</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    as.matrix(SummarizedExperiment::assay(se))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">137</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">139</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">get_tx2gene_from_se &lt;- function(se, readcounts_mat = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">140</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    md &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">141</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    try(md &lt;- S4Vectors::metadata(se), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">142</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # prefer explicit tx2gene in metadata</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">143</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(md) &amp;&amp; !is.null(md$tx2gene) &amp;&amp; is.data.frame(md$tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">144</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txmap &lt;- md$tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">145</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # attempt to find sensible columns</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tx_col &lt;- if ("Transcript" %in% colnames(txmap)) "Transcript" else colnames(txmap)[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">147</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        gene_col &lt;- if ("Gen" %in% colnames(txmap)) "Gen" else colnames(txmap)[2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">148</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">149</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            type = "vector",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            mapping = as.character(txmap[[gene_col]][match(rownames(readcounts_mat), txmap[[tx_col]])])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">151</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">152</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">153</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">154</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # fallback: try rowData mapping</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">155</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rdata &lt;- SummarizedExperiment::rowData(se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">156</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(rdata) &amp;&amp; "genes" %in% colnames(rdata)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">157</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        genes_vec &lt;- as.character(rdata$genes)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">158</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(readcounts_mat) &amp;&amp; length(genes_vec) == nrow(readcounts_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">159</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(list(type = "vector", mapping = genes_vec))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">161</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">162</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">163</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # last resort: use rownames of readcounts as gene identifiers</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">164</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(readcounts_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">165</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(type = "vector", mapping = rownames(readcounts_mat)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">166</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">167</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">168</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">169</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">170</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">171</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">validate_control_in_samples &lt;- function(control, samples) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">172</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uniq &lt;- unique(samples)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">173</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(control) &amp;&amp; control %in% uniq) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">174</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(control)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">175</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">176</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if ("Normal" %in% uniq) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">177</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return("Normal")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">178</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">179</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # fallback to first level and message</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">180</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    chosen &lt;- uniq[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">181</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    message(sprintf("`control` not found; using '%s' instead", chosen))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">182</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    chosen</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">183</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">184</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">185</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot q-curve profile for a single gene comparing groups</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">186</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">187</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' For a selected gene, plot per-sample Tsallis entropy across q values and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">188</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' overlay per-group median +/- IQR ribbons so group-level differences are</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">189</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' easy to compare. Expects a `SummarizedExperiment` produced by</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">190</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `calculate_diversity()` with `_q=` suffixes in column names.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">191</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">192</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` from `calculate_diversity()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">193</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param gene Character scalar or vector; gene symbol(s) to plot. If NULL and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">194</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `lm_res` is supplied, the top `n_top` genes from `lm_res` (by</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">195</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `adj_p_interaction` or `p_interaction`) are used.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">196</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param lm_res Optional data.frame result from `calculate_lm_interaction()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">197</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   When supplied and `gene` is NULL, the top `n_top` significant genes will</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">198</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   be plotted.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">199</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param n_top Number of top genes to plot when `lm_res` is provided (default: 10).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">200</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">201</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Column name in `colData(se)` with sample type labels</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">202</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   (default: "sample_type"). If missing, a single-group fallback is used.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">203</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param show_samples Logical; if TRUE, draw per-sample lines in the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">204</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   background (default: FALSE).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">205</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` object when a single gene is requested, or a named list</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">206</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   of `ggplot` objects when multiple genes are requested.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">207</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">208</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' mat &lt;- matrix(runif(8), nrow = 2, dimnames = list(c("g1", "g2"), c("s1_q=0.1", "s1_q=1", "s2_q=0.1", "s2_q=1")))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">209</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- SummarizedExperiment::SummarizedExperiment(assays = list(diversity = mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">210</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_tsallis_gene_profile(se, gene = "g1")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">211</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">212</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_tsallis_gene_profile &lt;- function(se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">213</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      gene = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">214</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      lm_res = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">215</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      n_top = 10,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">216</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      assay_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">217</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      sample_type_col = "sample_type",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">218</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                                      show_samples = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">219</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("ggplot2", quietly = TRUE)) stop("ggplot2 required")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">220</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("dplyr", "tidyr", "SummarizedExperiment"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">221</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">222</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- prepare_tsallis_long(se, assay_name = assay_name, sample_type_col = sample_type_col)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">223</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!("Gene" %in% colnames(long))) stop("prepare_tsallis_long did not return Gene column")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">224</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">225</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Resolve genes to plot: accept NULL (use lm_res), a single name, or a vector/list</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">226</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">227</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(lm_res)) stop("Either 'gene' or 'lm_res' must be provided")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">228</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.data.frame(lm_res) || !("gene" %in% colnames(lm_res))) stop("'lm_res' must be a data.frame with a 'gene' column")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">229</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # prefer adj_p_interaction if present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">230</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pcol &lt;- if ("adj_p_interaction" %in% colnames(lm_res)) "adj_p_interaction" else if ("p_interaction" %in% colnames(lm_res)) "p_interaction" else NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">231</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(pcol)) stop("'lm_res' must contain 'adj_p_interaction' or 'p_interaction' columns")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">232</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        genes_ordered &lt;- unique(as.character(lm_res$gene[order(lm_res[[pcol]])]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">233</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        genes &lt;- head(genes_ordered, n_top)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">234</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">235</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        genes &lt;- as.character(unlist(gene))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">236</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">237</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">238</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(genes) == 0) stop("No genes selected for plotting")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">239</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">240</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # helper to build single plot for a gene</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">241</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    make_plot_for_gene &lt;- function(sel) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">242</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long_g &lt;- long[as.character(long$Gene) == sel, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">243</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (nrow(long_g) == 0) stop("Gene not found in assay: ", sel)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">244</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long_g$qnum &lt;- as.numeric(as.character(long_g$q))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">245</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stats_df &lt;- dplyr::summarise(dplyr::group_by(long_g, group, qnum),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">246</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            median = median(tsallis, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">247</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            IQR = stats::IQR(tsallis, na.rm = TRUE), .groups = "drop"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">248</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">249</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">250</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p &lt;- ggplot2::ggplot() +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">251</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::theme_minimal(base_size = 14)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">252</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">253</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (isTRUE(show_samples)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">254</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p &lt;- p + ggplot2::geom_line(data = long_g, ggplot2::aes(x = qnum, y = tsallis, group = sample, color = group), alpha = 0.25)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">255</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">256</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">257</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p &lt;- p +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">258</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::geom_ribbon(data = stats_df, ggplot2::aes(x = qnum, ymin = median - IQR / 2, ymax = median + IQR / 2, fill = group), alpha = 0.2, inherit.aes = FALSE) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">259</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::geom_line(data = stats_df, ggplot2::aes(x = qnum, y = median, color = group), linewidth = 1.3) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">260</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::labs(title = paste0(sel, ": Tsallis entropy q-curve profile"), x = "q value", y = "Tsallis entropy", color = "Group", fill = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">261</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::scale_color_discrete(name = "Group") + ggplot2::scale_fill_discrete(name = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">262</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::theme(plot.title = ggplot2::element_text(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">263</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                hjust = 0.5, size = 16,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">264</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                margin = ggplot2::margin(b = 10)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">265</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">266</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">267</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">268</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">269</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Return single ggplot for single gene, or a named list of ggplots for multiple genes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">270</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(genes) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">271</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(make_plot_for_gene(genes))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">272</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">273</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">274</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plots &lt;- lapply(genes, make_plot_for_gene)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">275</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    names(plots) &lt;- genes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">276</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plots</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">277</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">278</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">279</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot diversity distributions (density) by sample type</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">280</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` returned by `calculate_diversity`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">281</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">282</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Optional column name in `colData(se)` with sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">283</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' types.  If missing, sample types are inferred from column names (suffix after</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">284</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the last underscore) or set to 'Group'.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">285</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` object with layered density plots.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">286</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom ggplot2 ggplot aes geom_density facet_grid scale_color_manual</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">287</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' guides theme_minimal labs</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">288</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">289</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">290</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data("tcga_brca_luma_dataset", package = "TSENAT")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">291</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">292</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">293</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = 0.1, norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">294</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_diversity_density(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">295</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_diversity_density &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">296</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">297</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  assay_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">298</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sample_type_col = NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">299</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">300</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2", "tidyr", "dplyr", "SummarizedExperiment"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">301</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- get_assay_long(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">302</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        se,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">303</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assay_name = assay_name,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">304</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        value_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">305</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_type_col = sample_type_col</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">306</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">307</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">308</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">309</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">310</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::aes(x = diversity, group = sample, color = sample_type)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">311</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">312</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_density(alpha = 0.3) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">313</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::facet_grid(. ~ sample_type) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">314</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_color_manual(values = c("black", "darkorchid4")) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">315</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::guides(color = "none") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">316</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">317</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(x = "Diversity values", y = "Density")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">318</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">319</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">320</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">321</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot violin of per-gene mean diversity by sample type</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">322</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom magrittr %&gt;%</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">323</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` returned by `calculate_diversity`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">324</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">325</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Optional column name in `colData(se)` containing</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">326</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' sample types.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">327</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` violin plot object.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">328</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">329</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">330</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data("tcga_brca_luma_dataset", package = "TSENAT")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">331</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">332</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">333</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = 0.1, norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">334</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_mean_violin(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">335</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_mean_violin &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">336</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">337</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  assay_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">338</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sample_type_col = NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">339</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">340</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2", "dplyr", "SummarizedExperiment", "tidyr"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">341</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- get_assay_long(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">342</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        se,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">343</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assay_name = assay_name,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">344</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        value_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">345</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_type_col = sample_type_col</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">346</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">347</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">348</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tmp &lt;- as.data.frame(long)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">349</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plot_df &lt;- stats::aggregate(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">350</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        diversity ~ sample_type + Gene,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">351</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        data = tmp,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">352</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        FUN = function(x) mean(x, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">353</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">354</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(plot_df)[colnames(plot_df) == "diversity"] &lt;- "value"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">355</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">356</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">357</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        plot_df,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">358</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::aes(x = sample_type, y = value, fill = sample_type)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">359</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">360</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_violin(alpha = 0.6) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">361</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::coord_flip() +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">362</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">363</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(x = "Samples", y = "Diversity") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">364</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_fill_viridis_d(name = "Group")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">365</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">366</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">367</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">368</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Core MA plotting implementation documentation moved to internal block</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">369</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot MA using Tsallis-based fold changes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">370</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">371</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Wrapper around `plot_ma(..., type = "tsallis")` for convenience and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">372</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' clearer API separation.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">373</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">374</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x Data.frame from `calculate_difference()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">375</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sig_alpha Numeric significance threshold for adjusted p-values (default: 0.05).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">376</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x_label Optional x-axis label passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">377</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param y_label Optional y-axis label passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">378</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param title Optional plot title passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">379</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param ... Additional arguments passed to `plot_ma()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">380</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot2` object representing the MA plot.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">381</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">382</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' x &lt;- data.frame(genes = paste0("g", seq_len(5)), mean = runif(5), log2_fold_change = rnorm(5))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">383</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_ma_tsallis(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">384</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">385</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_ma_tsallis &lt;- function(x, sig_alpha = 0.05, x_label = NULL, y_label = NULL, title = NULL, ...) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">386</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    title_use &lt;- title %||% "Tsallis-based MA plot"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">387</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_label_use &lt;- x_label %||% "mean_difference"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">388</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    y_label_use &lt;- y_label %||% "Log10 fold-change of entropy"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">389</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .plot_ma_core(x, fc_df = NULL, sig_alpha = sig_alpha, x_label = x_label_use, y_label = y_label_use, title = title_use)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">390</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">391</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">392</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">393</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot MA using expression/readcount-based fold changes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">394</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">395</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Wrapper around `plot_ma(..., type = "expression")` that accepts a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">396</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `SummarizedExperiment` or precomputed fold-change `data.frame`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">397</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">398</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x Data.frame from `calculate_difference()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">399</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` or data.frame supplying readcounts or precomputed fold changes.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">400</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Optional sample grouping vector (passed to `plot_ma`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">401</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param control Control level name (passed to `plot_ma`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">402</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param fc_method Aggregation method for fold-change calculation (passed to `plot_ma`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">403</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pseudocount Pseudocount added when computing log ratios (passed to `plot_ma`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">404</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sig_alpha Numeric significance threshold for adjusted p-values (default: 0.05).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">405</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x_label Optional x-axis label passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">406</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param y_label Optional y-axis label passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">407</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param title Optional plot title passed to `plot_ma`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">408</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param ... Additional arguments passed to `plot_ma()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">409</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot2` object representing the MA plot.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">410</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">411</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' x &lt;- data.frame(genes = paste0("g", seq_len(5)), mean = runif(5))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">412</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' fc &lt;- data.frame(genes = paste0("g", seq_len(5)), log2_fold_change = rnorm(5))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">413</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_ma_expression(x, se = fc)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">414</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">415</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_ma_expression &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">416</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  x,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">417</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">418</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  samples = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">419</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  control = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">420</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  fc_method = "median",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">421</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  pseudocount = 0,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">422</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sig_alpha = 0.05,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">423</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  x_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">424</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  y_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">425</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  title = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">426</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  ...</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">427</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">428</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    title_use &lt;- title %||% "Readcounts-based MA plot"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">429</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_label_use &lt;- x_label %||% "Mean difference"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">430</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    y_label_use &lt;- y_label %||% "Log10 fold-change of counts"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">431</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plot_ma_expression_impl(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">432</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">433</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        se = se,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">434</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        samples = samples,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">435</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        control = control,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">436</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fc_method = fc_method,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">437</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pseudocount = pseudocount,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">438</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sig_alpha = sig_alpha,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">439</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label = x_label_use,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">440</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        y_label = y_label_use,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">441</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        title = title_use,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">442</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ...</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">443</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">444</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">445</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">446</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">447</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Core MA plotting implementation used by wrappers above. Accepts a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">448</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># differential results `x` (data.frame) and an optional `fc_df` with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">449</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># fold-changes (genes as rownames or a `genes` column). Returns a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">450</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># `ggplot` MA-plot.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">451</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Core MA plotting implementation (internal)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">452</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">453</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' This is an internal helper used by `plot_ma_tsallis()` and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">454</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `plot_ma_expression()`. It is documented here for developers but</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">455</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' is not exported.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">456</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @noRd</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">457</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.plot_ma_core &lt;- function(x,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">458</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          fc_df = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">459</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          diff_res = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">460</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          sig_alpha = 0.05,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">461</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          x_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">462</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          y_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">463</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          title = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">464</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                          ...) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">465</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">466</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">467</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- as.data.frame(x, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">468</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # ensure a gene identifier column exists</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">469</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!("genes" %in% colnames(df))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">470</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(rownames(df))) df$genes &lt;- rownames(df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">471</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">472</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">473</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If an external fc_df is provided, merge fold values into df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">474</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(fc_df)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">475</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fdf &lt;- as.data.frame(fc_df, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">476</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!("genes" %in% colnames(fdf)) &amp;&amp; !is.null(rownames(fdf))) fdf$genes &lt;- rownames(fdf)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">477</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!("log2_fold_change" %in% colnames(fdf))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">478</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Provided `fc_df` must contain 'log2_fold_change' column")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">479</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">480</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df &lt;- merge(df, fdf[, c("genes", "log2_fold_change")], by = "genes", all.x = TRUE, suffixes = c("", ".fc"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">481</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # prefer fc_df fold values when present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">482</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if ("log2_fold_change.fc" %in% colnames(df)) df$log2_fold_change &lt;- ifelse(!is.na(df$log2_fold_change.fc), df$log2_fold_change.fc, df$log2_fold_change)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">483</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">484</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">485</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Detect fold-change column</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">486</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fold_candidates &lt;- c("log2_fold_change", "logFC", "fold", "estimate_interaction", "fold_change")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">487</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fold_col &lt;- intersect(fold_candidates, colnames(df))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">488</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(fold_col) == 0) stop("Could not find a fold-change column in input")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">489</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fold_col &lt;- fold_col[1]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">490</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">491</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Detect mean/average columns for x-axis</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">492</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mean_cols &lt;- grep("_mean$|_median$", colnames(df), value = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">493</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(mean_cols) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">494</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # ensure the two chosen columns are consistent (both _mean or both _median)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">495</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        c1 &lt;- mean_cols[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">496</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        c2 &lt;- mean_cols[2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">497</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        is_mean1 &lt;- grepl("_mean$", c1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">498</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        is_mean2 &lt;- grepl("_mean$", c2)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">499</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        is_med1 &lt;- grepl("_median$", c1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">500</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        is_med2 &lt;- grepl("_median$", c2)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">501</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!((is_mean1 &amp;&amp; is_mean2) || (is_med1 &amp;&amp; is_med2))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">502</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Could not find two mean or two median columns")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">503</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">504</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- rowMeans(df[, mean_cols[seq_len(2)],</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">505</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            drop = FALSE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">506</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ], na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">507</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% paste0(mean_cols[1], " vs ", mean_cols[2])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">508</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (length(mean_cols) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">509</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- as.numeric(df[[mean_cols[1]]])</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">510</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% mean_cols[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">511</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if ("mean" %in% colnames(df)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">512</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- as.numeric(df$mean)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">513</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% "Mean"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">514</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">515</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # fallback: use rank or index</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">516</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- seq_len(nrow(df))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">517</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% "Index"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">518</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">519</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">520</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    yvals &lt;- as.numeric(df[[fold_col]])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">521</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">522</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # p-value / adjusted p-value detection</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">523</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_candidates &lt;- c("adjusted_p_values", "adj_p_value", "adj_p", "padj", "p.adjust")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">524</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_col &lt;- intersect(padj_candidates, colnames(df))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">525</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_col &lt;- if (length(padj_col)) padj_col[1] else NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">526</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">527</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj &lt;- if (!is.null(padj_col)) as.numeric(df[[padj_col]]) else rep(1, length(yvals))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">528</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj[is.na(padj)] &lt;- 1</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">529</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">530</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sig_flag &lt;- ifelse(abs(yvals) &gt; 0 &amp; padj &lt; sig_alpha, "significant", "non-significant")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">531</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">532</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plot_df &lt;- data.frame(genes = df$genes, x = xvals, y = yvals, padj = padj, significant = sig_flag, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">533</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">534</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    prep &lt;- .tsenat_prepare_ma_plot_df(df, fold_col = fold_col, mean_cols = mean_cols, x_label = x_label, y_label = y_label)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">535</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plot_df &lt;- prep$plot_df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">536</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_label_formatted &lt;- .tsenat_format_label(prep$x_label)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">537</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    y_label_raw &lt;- prep$y_label %||% fold_col</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">538</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    y_label_formatted &lt;- .tsenat_format_label(y_label_raw)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">539</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(y_label_formatted)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">540</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        y_label_formatted &lt;- sub("\\blog2\\b", "log10", y_label_formatted, ignore.case = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">541</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">542</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">543</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p &lt;- ggplot2::ggplot(plot_df, ggplot2::aes(x = x, y = y, color = significant)) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">544</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_point(alpha = 0.75, size = 3.2) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">545</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_color_manual(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">546</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            values = c("non-significant" = "grey40", "significant" = "firebrick3"),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">547</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            guide = "none"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">548</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">549</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">550</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">551</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            title = title %||% "MA plot: mean vs log10 fold-change",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">552</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x = x_label_formatted,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">553</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            y = y_label_formatted</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">554</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">555</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">556</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            plot.title = ggplot2::element_text(hjust = 0.5, size = 16, face = "plain"),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">557</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            axis.title = ggplot2::element_text(size = 14),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">558</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            axis.text = ggplot2::element_text(size = 12)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">559</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">560</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">561</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">562</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">563</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">564</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">565</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Implementation that computes fold-changes from expression/readcounts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">566</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># stored in a `SummarizedExperiment` (or matrix) and forwards to core plotter.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">567</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_ma_expression_impl &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">568</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  x,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">569</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">570</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  samples = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">571</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  control = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">572</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  fc_method = "median",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">573</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  pseudocount = 0,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">574</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sig_alpha = 0.05,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">575</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  x_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">576</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  y_label = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">577</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  title = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">578</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  ...</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">579</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">580</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("SummarizedExperiment"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">581</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">582</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # extract/readcounts</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">583</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(se, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">584</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        counts &lt;- get_readcounts_from_se(se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">585</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        samples &lt;- infer_samples_from_se(se, samples)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">586</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(samples)) stop("Could not infer 'samples' from SummarizedExperiment; provide `samples`")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">587</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        control &lt;- validate_control_in_samples(control, samples)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">588</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">589</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # attempt to map transcripts -&gt; genes and aggregate counts per gene</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">590</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tx2g &lt;- get_tx2gene_from_se(se, readcounts_mat = counts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">591</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(tx2g) &amp;&amp; tx2g$type == "vector") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">592</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            mapping &lt;- tx2g$mapping</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">593</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # ensure mapping length matches rows</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">594</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(mapping) == nrow(counts)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">595</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # aggregate transcript-level counts to gene-level using rowsum</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">596</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                agg &lt;- rowsum(counts, group = mapping)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">597</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                counts_gene &lt;- as.matrix(agg)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">598</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">599</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                counts_gene &lt;- counts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">600</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">601</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">602</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            counts_gene &lt;- counts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">603</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">604</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">605</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # compute fold-changes using calculate_fc (aggregates per-group)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">606</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fc_res &lt;- calculate_fc(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">607</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            counts_gene,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">608</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            samples,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">609</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            control,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">610</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            method = fc_method,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">611</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pseudocount = pseudocount</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">612</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">613</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # ensure genes column exists</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">614</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(rownames(fc_res))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">615</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fc_res$genes &lt;- seq_len(nrow(fc_res))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">616</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">617</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fc_res$genes &lt;- rownames(fc_res)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">618</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">619</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(.plot_ma_core(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">620</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">621</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fc_df = fc_res,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">622</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            sig_alpha = sig_alpha,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">623</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x_label = x_label,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">624</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            y_label = y_label,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">625</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            title = title,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">626</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ...</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">627</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">628</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">629</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">630</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If se is provided as a matrix/data.frame of precomputed fold changes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">631</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.matrix(se) || is.data.frame(se)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">632</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fc_res &lt;- as.data.frame(se, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">633</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!("log2_fold_change" %in% colnames(fc_res))) stop("`se` data.frame must contain 'log2_fold_change' column when providing precomputed fold changes")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">634</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!("genes" %in% colnames(fc_res)) &amp;&amp; !is.null(rownames(fc_res))) fc_res$genes &lt;- rownames(fc_res)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">635</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(.plot_ma_core(x, fc_df = fc_res, sig_alpha = sig_alpha, x_label = x_label, y_label = y_label, title = title, ...))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">636</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">637</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">638</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    stop("Unsupported 'se' argument for plot_ma_expression_impl")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">639</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">640</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">641</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">642</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot median +- IQR of Tsallis entropy across q values by group</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">643</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">644</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' This reproduces the `tsallis-q-curve-mean-sd` plot from the vignette: for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">645</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' each q value, compute per-gene Tsallis entropy per sample, summarize across</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">646</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' genes by group (median and IQR) and plot median with a ribbon spanning</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">647</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' median +- IQR/2.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">648</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` returned by `calculate_diversity`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">649</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">650</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Column name in `colData(se)` containing sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">651</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   type labels (default: "sample_type").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">652</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` object showing median +- IQR across q values by group.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">653</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">654</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">655</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data("tcga_brca_luma_dataset", package = "TSENAT")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">656</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:40, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">657</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:40]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">658</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">659</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     q = seq(0.01, 0.1, by = 0.03), norm = TRUE</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">660</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">661</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' p &lt;- plot_tsallis_q_curve(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">662</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">663</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_tsallis_q_curve &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">664</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  se,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">665</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  assay_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">666</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sample_type_col = "sample_type"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">667</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">668</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # SE-first API: require a SummarizedExperiment with per-column sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">669</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # type mapping in `colData(se)[, sample_type_col]` (or allow a single</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">670</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # group dataset where `sample_type` is omitted).</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">671</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(se, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">672</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        require_pkgs(c("ggplot2", "dplyr", "tidyr", "SummarizedExperiment"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">673</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long &lt;- prepare_tsallis_long(se, assay_name = assay_name, sample_type_col = sample_type_col)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">674</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        y_label &lt;- "Tsallis entropy (S_q)"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">675</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (nrow(long) == 0) stop("No tsallis values found in SummarizedExperiment")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">676</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # long$q may be a factor; convert to numeric for plotting</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">677</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$qnum &lt;- as.numeric(as.character(long$q))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">678</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stats_df &lt;- dplyr::summarise(dplyr::group_by(long, group, qnum),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">679</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            median = median(tsallis, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">680</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            IQR = stats::IQR(tsallis, na.rm = TRUE), .groups = "drop"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">681</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">682</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p &lt;- ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">683</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stats_df,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">684</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::aes(x = qnum, y = median, color = group, fill = group)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">685</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">686</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::geom_line(linewidth = 1.3) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">687</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::geom_ribbon(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">688</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                ggplot2::aes(ymin = median - IQR / 2, ymax = median + IQR / 2),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">689</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                alpha = 0.2,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">690</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                color = NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">691</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">692</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">693</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::labs(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">694</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                title = "Global Tsallis q-curve: entropy across all genes",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">695</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                x = "q value",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">696</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                y = y_label,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">697</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                color = "Group",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">698</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                fill = "Group"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">699</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">700</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, face = "plain", size = 16))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">701</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # use default discrete ggplot2 colours (not viridis)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">702</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p &lt;- p + ggplot2::scale_color_discrete(name = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">703</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ggplot2::scale_fill_discrete(name = "Group")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">704</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # If there is only a single group present, hide the legend/Group label</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">705</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(unique(stats_df$group)) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">706</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p &lt;- p + ggplot2::theme(legend.position = "none")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">707</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">708</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">709</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(p)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">710</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">711</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">712</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Matrix/data.frame input is no longer supported for this plot function.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">713</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    stop(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">714</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "plot_tsallis_q_curve requires a SummarizedExperiment from calculate_diversity."</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">715</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">716</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">717</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Violin plot of Tsallis entropy for multiple q values</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">718</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">719</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` returned by `calculate_diversity` with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">720</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' multiple q values (column names contain `_q=`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">721</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">722</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` violin plot object faceted/colored by group and q.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">723</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">724</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">725</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data("tcga_brca_luma_dataset", package = "TSENAT")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">726</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">727</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">728</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = c(0.1, 1), norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">729</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_tsallis_violin_multq(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">730</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_tsallis_violin_multq &lt;- function(se, assay_name = "diversity") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">731</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2", "tidyr", "dplyr"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">732</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- prepare_tsallis_long(se, assay_name = assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">733</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">734</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">735</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">736</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::aes(x = q, y = tsallis, fill = group)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">737</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">738</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_violin(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">739</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            alpha = 0.5, width = 0.9,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">740</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            position = ggplot2::position_dodge(width = 0.8)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">741</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">742</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_boxplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">743</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            width = 0.15,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">744</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            position = ggplot2::position_dodge(width = 0.8),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">745</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            outlier.shape = NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">746</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">747</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">748</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_fill_discrete(name = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">749</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">750</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            title = "Violin plot: Tsallis entropy distribution across multiple q values",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">751</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x = "q value",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">752</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            y = "Tsallis entropy",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">753</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fill = "Group"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">754</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">755</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(plot.title = ggplot2::element_text(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">756</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            hjust = 0.5, size = 16, face = "plain",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">757</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            margin = ggplot2::margin(b = 10)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">758</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">759</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">760</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Density plot of Tsallis entropy for multiple q values</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">761</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">762</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` returned by `calculate_diversity` with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">763</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' multiple q values (column names contain `_q=`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">764</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay to use (default: "diversity").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">765</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot` density plot object faceted by q and colored by group.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">766</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">767</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">768</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data("tcga_brca_luma_dataset", package = "TSENAT")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">769</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">770</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">771</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = c(0.1, 1), norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">772</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_tsallis_density_multq(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">773</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_tsallis_density_multq &lt;- function(se, assay_name = "diversity") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">774</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2", "tidyr", "dplyr"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">775</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- prepare_tsallis_long(se, assay_name = assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">776</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">777</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">778</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">779</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::aes(x = tsallis, color = group, fill = group)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">780</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">781</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_density(alpha = 0.3) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">782</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::facet_wrap(~q, scales = "free_y") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">783</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">784</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_color_discrete(name = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">785</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_fill_discrete(name = "Group") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">786</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">787</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            title = "Density plot: Tsallis entropy distribution across multiple q values",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">788</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x = "Tsallis entropy",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">789</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            y = "Density",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">790</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            color = "Group",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">791</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fill = "Group"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">792</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">793</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(plot.title = ggplot2::element_text(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">794</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            hjust = 0.5, size = 16, face = "plain",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">795</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            margin = ggplot2::margin(b = 10)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">796</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">797</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">798</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">799</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">800</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Volcano plot for differential results</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">801</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">802</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Create a volcano plot showing fold-change (x-axis) versus adjusted</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">803</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' p-value significance (y-axis). The function auto-detects a suitable x-axis</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">804</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' column if one is not provided and expects an adjusted p-value column for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">805</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' significance coloring.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">806</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">807</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param diff_df Data.frame from `test_differential()` or similar results.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">808</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   Should contain p-values and optionally fold-change columns.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">809</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x_col Optional column name for the x-axis. If `NULL`, the function</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">810</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   will try to auto-detect a suitable numeric column (excluding p-values).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">811</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param padj_col Adjusted p-value column name (default: "adjusted_p_values").</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">812</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param label_thresh Fold-change threshold used to annotate points (default: 0.1).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">813</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param padj_thresh Adjusted p-value cutoff for significance (default: 0.05).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">814</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param top_n Number of top significant genes to label (default: 5).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">815</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param title Optional plot title; if `NULL` a default title is used.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">816</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">817</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A `ggplot2` object.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">818</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">819</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">820</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' df &lt;- data.frame(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">821</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     gene = paste0("g", seq_len(10)),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">822</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     mean_difference = runif(10),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">823</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     adjusted_p_values = runif(10)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">824</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">825</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' # plot_volcano(df, x_col = "mean_difference", padj_col = "adjusted_p_values")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">826</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_volcano &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">827</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  diff_df,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">828</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  x_col = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">829</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  padj_col = "adjusted_p_values",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">830</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  label_thresh = 0.1,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">831</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  padj_thresh = 0.05,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">832</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  top_n = 5,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">833</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  title = NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">834</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">835</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("ggplot2", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">836</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("ggplot2 required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">837</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">838</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">839</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    prep_volcano &lt;- .tsenat_prepare_volcano_df(diff_df = diff_df, x_col = x_col, padj_col = padj_col, label_thresh = label_thresh, padj_thresh = padj_thresh, title = title)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">840</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- prep_volcano$df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">841</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_col &lt;- prep_volcano$x_col</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">842</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_col &lt;- prep_volcano$padj_col</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">843</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_label_formatted &lt;- prep_volcano$x_label_formatted</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">844</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_label_formatted &lt;- prep_volcano$padj_label_formatted</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">845</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    title_use &lt;- prep_volcano$title_use</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">846</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">847</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p &lt;- ggplot2::ggplot(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">848</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">849</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::aes(x = xval, y = -log10(padj), color = significant)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">850</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">851</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_point(alpha = 0.75, size = 3.4) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">852</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::scale_color_manual(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">853</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            values = c("non-significant" = "black", "significant" = "red"),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">854</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            guide = "none"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">855</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">856</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_hline(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">857</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            yintercept = -log10(padj_thresh),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">858</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            linetype = "dashed",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">859</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            color = "gray50"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">860</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">861</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_vline(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">862</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            xintercept = c(-label_thresh, label_thresh),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">863</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            linetype = "dashed",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">864</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            color = "gray50"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">865</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">866</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">867</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">868</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            title = title_use,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">869</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x = x_label_formatted,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">870</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            y = paste0("-Log10(", padj_label_formatted, ")")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">871</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">872</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">873</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            plot.title = ggplot2::element_text(hjust = 0.5, size = 16, face = "plain", margin = ggplot2::margin(b = 4)),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">874</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            axis.title = ggplot2::element_text(size = 14),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">875</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            axis.text = ggplot2::element_text(size = 12)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">876</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">877</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">878</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">879</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">880</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">881</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">882</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Internal helper to compute fill limits across multiple genes (not exported)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">883</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @noRd</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">884</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.compute_transcript_fill_limits &lt;- function(genes, mapping, counts, samples, top_n, agg_fun, pseudocount) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">885</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mins &lt;- maxs &lt;- c()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">886</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (g in genes) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">887</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txs &lt;- mapping$Transcript[mapping$Gen == g]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">888</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txs &lt;- intersect(txs, rownames(counts))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">889</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(txs) == 0) next</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">890</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(top_n)) txs &lt;- head(txs, top_n)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">891</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mat &lt;- counts[txs, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">892</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_all &lt;- as.data.frame(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">893</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_all$tx &lt;- rownames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">894</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_long &lt;- tidyr::pivot_longer(df_all, -tx, names_to = "sample", values_to = "expr")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">895</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_long$group &lt;- rep(samples, times = length(txs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">896</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_summary &lt;- stats::aggregate(expr ~ tx + group, data = df_long, FUN = agg_fun)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">897</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_summary$log2expr &lt;- log2(df_summary$expr + pseudocount)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">898</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mins &lt;- c(mins, min(df_summary$log2expr, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">899</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        maxs &lt;- c(maxs, max(df_summary$log2expr, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">900</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">901</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(mins) == 0) stop("No transcripts found for provided genes")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">902</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    c(min(mins, na.rm = TRUE), max(maxs, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">903</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">904</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">905</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Internal helper to draw grid layout with title, plots, and legend using base grid</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">906</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @noRd</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">907</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.draw_transcript_grid &lt;- function(grobs, title, legend_grob, n, heights, to_file = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">908</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If no output file is provided and no graphics device is open, render to a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">909</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # temporary pdf device so that plotting in non-interactive sessions does not</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">910</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # create `Rplots.pdf` in the working directory.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">911</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    temp_dev &lt;- FALSE</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">912</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Only open a temporary PDF device when:</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">913</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    #  - caller did not supply an output file (`to_file` is NULL),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">914</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    #  - the session is non-interactive, and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">915</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    #  - no graphics device is currently open (dev.cur() == 1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">916</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(to_file) &amp;&amp; !interactive() &amp;&amp; grDevices::dev.cur() == 1L) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">917</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tmp &lt;- tempfile("TSENAT_plot_", fileext = ".pdf")</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">918</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grDevices::pdf(tmp)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">919</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        temp_dev &lt;- TRUE</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">920</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Ensure device is closed and temporary file removed on exit</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">921</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        on.exit(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">922</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">923</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                try(grDevices::dev.off(), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">924</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (file.exists(tmp)) unlink(tmp)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">925</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            },</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">926</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            add = TRUE</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">927</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">928</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">929</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">930</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::grid.newpage()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">931</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::pushViewport(</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">932</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::viewport(layout = grid::grid.layout(3, n, heights = heights))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">933</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">934</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Title row</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">935</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    vp_title &lt;- grid::viewport(layout.pos.row = 1, layout.pos.col = seq_len(n))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">936</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::pushViewport(vp_title)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">937</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::grid.text(title, x = 0.5, gp = grid::gpar(fontsize = 12))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">938</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::upViewport()</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">939</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Plot rows</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">940</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (i in seq_along(grobs)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">941</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        vp &lt;- grid::viewport(layout.pos.row = 2, layout.pos.col = i)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">942</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::pushViewport(vp)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">943</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::grid.draw(grobs[[i]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">944</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::upViewport()</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">945</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">946</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Legend row</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">947</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(legend_grob)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">948</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        vp_leg &lt;- grid::viewport(layout.pos.row = 3, layout.pos.col = seq_len(n))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">949</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::pushViewport(vp_leg)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">950</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::grid.draw(legend_grob)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">951</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        grid::upViewport()</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">952</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">953</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid::upViewport()</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">954</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If caller supplied a file (caller likely opened a device), close it here.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">955</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(to_file)) grDevices::dev.off()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">956</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    invisible(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">957</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">958</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">959</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Internal helpers for `plot_top_transcripts` refactor</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">960</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Create per-gene plot and combine multiple gene plots into final output</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">961</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_make_plot_for_gene &lt;- function(gene_single, mapping, counts, samples, top_n, agg_fun, pseudocount, agg_label_unique, fill_limits = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">962</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2", "tidyr"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">963</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    built &lt;- .ptt_build_tx_long(gene_single, mapping, counts, samples, top_n)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">964</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_summary &lt;- .ptt_aggregate_df_long(built$df_long, agg_fun, pseudocount)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">965</td>
                    <td class="coverage">16<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .ptt_build_plot_from_summary(df_summary, agg_label_unique, fill_limits)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">966</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">967</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">968</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_combine_plots &lt;- function(plots, output_file = NULL, agg_label_unique = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">969</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    require_pkgs(c("ggplot2"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">970</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Allow callers to pass a single character second argument as the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">971</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # `agg_label_unique` for convenience (legacy test call patterns).</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">972</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(agg_label_unique) &amp;&amp; !is.null(output_file) &amp;&amp; is.character(output_file) &amp;&amp; length(output_file) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">973</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        agg_label_unique &lt;- output_file</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">974</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        output_file &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">975</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">976</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("patchwork", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">977</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .ptt_combine_patchwork(plots, agg_label_unique)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">978</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (requireNamespace("cowplot", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">979</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .ptt_combine_cowplot(plots, output_file = output_file, agg_label_unique = agg_label_unique)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">980</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">981</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .ptt_combine_grid(plots, output_file = output_file, agg_label_unique = agg_label_unique)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">982</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">983</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">984</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">985</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Prepare and validate inputs for `plot_top_transcripts`</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">986</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_prepare_inputs &lt;- function(counts, readcounts = NULL, samples = NULL, coldata = NULL, sample_type_col = "sample_type", tx2gene = NULL, res = NULL, top_n = NULL, pseudocount = 0, output_file = NULL, metric = c("median", "mean", "variance", "iqr")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">987</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # handle selecting genes from `res` is left to caller; this function focuses</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">988</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # on normalizing counts, samples and tx2gene mapping and preparing agg functions</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">989</td>
                    <td class="coverage">25<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(counts, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">990</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        require_pkgs(c("SummarizedExperiment", "S4Vectors"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">991</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        se &lt;- counts</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">992</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        counts_mat &lt;- get_readcounts_from_se(se, readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">993</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        counts &lt;- as.matrix(counts_mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">994</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        samples &lt;- infer_samples_from_se(se, samples, sample_type_col = sample_type_col)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">995</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">996</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">997</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            txres &lt;- get_tx2gene_from_se(se, counts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">998</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.null(txres) &amp;&amp; !is.null(txres$mapping)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">999</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                mapping &lt;- data.frame(Transcript = rownames(counts), Gen = as.character(txres$mapping), stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1000</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                tx2gene &lt;- mapping</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1001</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1002</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1003</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1004</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1005</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.matrix(counts) &amp;&amp; !is.data.frame(counts)) stop("`counts` must be a matrix or data.frame with transcripts as rownames")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1006</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    counts &lt;- as.matrix(counts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1007</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(rownames(counts))) stop("`counts` must have rownames corresponding to transcript identifiers")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1008</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1009</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # derive samples from coldata if needed</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1010</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(samples)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1011</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(coldata)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1012</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (is.character(coldata) &amp;&amp; length(coldata) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1013</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (!file.exists(coldata)) stop("coldata file not found: ", coldata)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1014</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                cdf &lt;- utils::read.delim(coldata, header = TRUE, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1015</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else if (is.data.frame(coldata)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1016</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                cdf &lt;- coldata</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1017</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1018</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("`coldata` must be a data.frame or path to a tab-delimited file")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1019</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1020</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1021</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.null(rownames(cdf)) &amp;&amp; all(colnames(counts) %in% rownames(cdf))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1022</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                samples &lt;- as.character(cdf[colnames(counts), sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1023</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1024</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                sample_id_cols &lt;- c("sample", "Sample", "sample_id", "id")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1025</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                sid &lt;- intersect(sample_id_cols, colnames(cdf))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1026</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (length(sid) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1027</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    sid &lt;- sid[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1028</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    if (!all(colnames(counts) %in% as.character(cdf[[sid]]))) stop("coldata sample id column does not match column names of counts")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1029</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    row_ix &lt;- match(colnames(counts), as.character(cdf[[sid]]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1030</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    samples &lt;- as.character(cdf[[sample_type_col]][row_ix])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1031</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1032</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    stop("Could not match `coldata` rows to `counts` columns. Provide `samples` or a row-named `coldata`.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1033</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1034</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1035</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1036</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Either 'samples' or 'coldata' must be provided to determine sample groups")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1037</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1038</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1039</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1040</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # normalize tx2gene mapping</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1041</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(tx2gene)) stop("`tx2gene` must be provided as a file path or data.frame (or include mapping in metadata of provided SummarizedExperiment)")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1042</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(tx2gene) &amp;&amp; length(tx2gene) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1043</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!file.exists(tx2gene)) stop("tx2gene file not found: ", tx2gene)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1044</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mapping &lt;- utils::read.delim(tx2gene, stringsAsFactors = FALSE, header = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1045</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.data.frame(tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1046</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mapping &lt;- tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1047</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1048</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`tx2gene` must be provided as a file path or data.frame (or include mapping in metadata of provided SummarizedExperiment)")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1049</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1050</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1051</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!all(c("Transcript", "Gen") %in% colnames(mapping))) stop("tx2gene must have columns 'Transcript' and 'Gen'")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1052</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">1053</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("ggplot2", quietly = TRUE)) stop("ggplot2 required for plotting")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1054</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1055</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(samples) &amp;&amp; length(samples) != ncol(counts)) stop("Length of `samples` must equal number of columns in `counts`")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1056</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1057</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    metric_choice &lt;- match.arg(metric)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1058</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_fun &lt;- switch(metric_choice,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1059</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        median = function(x) stats::median(x, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1060</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mean = function(x) base::mean(x, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1061</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        variance = function(x) stats::var(x, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1062</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        iqr = function(x) stats::IQR(x, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1063</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    )</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1064</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label_metric &lt;- if (metric_choice == "iqr") "IQR" else metric_choice</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1065</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label &lt;- sprintf("Transcript-level expression with metric %s", agg_label_metric)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1066</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .cnt &lt;- as.integer(getOption("TSENAT.plot_top_counter", 0)) + 1L</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1067</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    options(TSENAT.plot_top_counter = .cnt)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1068</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label_unique &lt;- agg_label</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1069</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1070</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(counts = counts, samples = samples, mapping = mapping, metric_choice = metric_choice, agg_fun = agg_fun, agg_label_unique = agg_label_unique, top_n = top_n, pseudocount = pseudocount, output_file = output_file)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1071</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1072</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1073</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Plot top transcripts for a gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1074</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param counts Matrix or data.frame of transcript counts. Rows are transcripts and columns are samples.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1075</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param readcounts Optional matrix or data.frame of raw read counts. Used for transcript-level quantification if provided.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1076</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param gene Character; gene symbol to inspect.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1077</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Character vector of sample group labels (length = ncol(counts)).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1078</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param coldata Optional data.frame or file path containing sample metadata. Used to infer sample groups if `samples` is not provided.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1079</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Character; column name in `coldata` or `SummarizedExperiment` colData to use for sample grouping. Default is "sample_type".</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1080</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param tx2gene Path or data.frame mapping transcripts to genes. Must contain columns `Transcript` and `Gen`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1081</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param res Optional result data.frame from a differential analysis. If provided and `gene` is NULL, top genes are selected by adjusted p-value.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1082</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param top_n Integer number of transcripts to show (default = 3). Use NULL to plot all transcripts for the gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1083</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pseudocount Numeric pseudocount added before log2 (default = 1e-6) to avoid division by zero.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1084</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param output_file Optional file path to save the plot. If `NULL`, the `ggplot` object is returned.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1085</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param metric Aggregation metric used to summarize transcript expression per group when plotting. One of c("median", "mean", "variance", "iqr"). Use "iqr" to compute the interquartile range. Defaults to "median".</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1086</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1087</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' tx_counts &lt;- matrix(sample(1:100, 24, replace = TRUE), nrow = 6)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1088</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rownames(tx_counts) &lt;- paste0("tx", seq_len(nrow(tx_counts)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1089</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' colnames(tx_counts) &lt;- paste0("S", seq_len(ncol(tx_counts)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1090</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' tx2gene &lt;- data.frame(Transcript = rownames(tx_counts), Gen = rep(paste0("G", seq_len(3)), each = 2), stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1091</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' samples &lt;- rep(c("Normal", "Tumor"), length.out = ncol(tx_counts))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1092</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' plot_top_transcripts(tx_counts, gene = c("G1", "G2"), samples = samples, tx2gene = tx2gene, top_n = 2)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1093</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1094</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">plot_top_transcripts &lt;- function(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1095</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  counts,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1096</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  readcounts = NULL, # Optional matrix or data.frame of raw read counts. Used for transcript-level quantification if provided.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1097</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  gene = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1098</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  samples = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1099</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  coldata = NULL, # Optional data.frame or file path containing sample metadata. Used to infer sample groups if `samples` is not provided.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  sample_type_col = "sample_type", # Column name in `coldata` or `SummarizedExperiment` colData to use for sample grouping. Default is "sample_type".</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  tx2gene = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1102</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  res = NULL, # Optional result data.frame from a differential analysis. If provided and `gene` is NULL, top genes are selected by adjusted p-value.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1103</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  top_n = 3,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1104</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  pseudocount = 1e-6,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1105</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  output_file = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">  metric = c("median", "mean", "variance", "iqr")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1107</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1108</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If `gene` is not provided, select top genes from `res` using `top_n`.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1109</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1110</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        gene &lt;- .ptt_select_genes_from_res(res, top_n)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1111</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1112</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    per_gene_top_n &lt;- top_n</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1113</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1114</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ## Prepare inputs and normalization via helper</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1115</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    prep &lt;- .ptt_prepare_inputs(counts = counts, readcounts = readcounts, samples = samples, coldata = coldata, sample_type_col = sample_type_col, tx2gene = tx2gene, res = res, top_n = per_gene_top_n, pseudocount = pseudocount, output_file = output_file, metric = metric)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1116</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1117</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # if prep returned without gene selection, caller will check `gene`</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1118</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    counts &lt;- prep$counts</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1119</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    samples &lt;- prep$samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1120</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mapping &lt;- prep$mapping</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1121</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    metric_choice &lt;- prep$metric_choice</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1122</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_fun &lt;- prep$agg_fun</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1123</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label_unique &lt;- prep$agg_label_unique</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1124</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    top_n &lt;- prep$top_n</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1125</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pseudocount &lt;- prep$pseudocount</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1126</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    output_file &lt;- prep$output_file</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1127</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1128</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    make_plot_for_gene &lt;- function(gene_single, fill_limits = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1129</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .ptt_make_plot_for_gene(gene_single, mapping, counts, samples, top_n, agg_fun, pseudocount, agg_label_unique, fill_limits)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1130</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1131</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1132</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Produce plots (single or multiple). Do not save inside helper - save once</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1133</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # below.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1134</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(gene) &gt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1135</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fill_limits &lt;- .compute_transcript_fill_limits(gene, mapping, counts, samples, top_n, agg_fun, pseudocount)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1136</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1137</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        plots &lt;- lapply(seq_along(gene), function(i) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1138</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            gname &lt;- gene[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1139</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pp &lt;- make_plot_for_gene(gname, fill_limits = fill_limits)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1140</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            per_gene_title &lt;- if (!is.na(gname) &amp;&amp; nzchar(as.character(gname))) as.character(gname) else ""</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1141</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pp &lt;- pp + ggplot2::labs(title = per_gene_title) + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, size = 16))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1142</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pp</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1143</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        })</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1144</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1145</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_plot &lt;- .ptt_combine_plots(plots, output_file = output_file, agg_label_unique = agg_label_unique)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1146</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1147</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_plot &lt;- make_plot_for_gene(gene)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1148</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1149</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1150</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(output_file)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1151</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::ggsave(output_file, result_plot)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1152</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        invisible(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1153</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">1154</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_plot</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1155</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">1156</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/map_functions.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Map external coldata into a SummarizedExperiment</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Map an external `coldata` table (sample IDs and a condition/label column)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' into a `SummarizedExperiment` produced by `calculate_diversity()`. The</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' helper sets `colData(ts_se)$sample_type` when possible and records a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `sample_base` identifier for each column. Unmatched entries remain NA and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' no automatic inference is attempted.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param ts_se A `SummarizedExperiment` whose assay columns are named with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   sample identifiers. Names may include per-sample annotations; mapping is</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   exact and case-sensitive unless you normalize identifiers beforehand.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param coldata A data.frame with sample metadata (or `NULL`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param coldata_sample_col Name of the column in `coldata` containing sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   identifiers (default: 'Sample').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param coldata_condition_col Name of the column in `coldata` with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   condition/labels (default: 'Condition').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired Logical; if `TRUE`, validate pairing and reorder columns so</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   that matched samples for each base are adjacent (default: `FALSE`). Use</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `paired = TRUE` for paired analyses so downstream paired tests see</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   aligned samples regardless of the original `coldata` order.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return The input `ts_se` with `colData(ts_se)$sample_type` populated when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' possible. `colData(ts_se)$sample_base` is also added containing the base</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' sample identifier; rownames of `colData(ts_se)` are aligned to the assay</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' column names.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data('tcga_brca_luma_dataset', package = 'TSENAT')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = 0.1, norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' sample_names &lt;- sub('_q=.*', '', colnames(SummarizedExperiment::assay(se)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">31</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' coldata_df &lt;- data.frame(Sample = sample_names, Condition = rep(c('A', 'B'),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     length.out = ncol(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' ))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' map_metadata(se, coldata_df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' # Optionally validate pairs when appropriate</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' map_metadata(se, coldata_df, paired = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">map_metadata &lt;- function(ts_se, coldata, coldata_sample_col = "Sample", coldata_condition_col = "Condition",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paired = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(coldata)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(ts_se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">42</td>
                    <td class="coverage">23<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.data.frame(coldata) || !all(c(coldata_sample_col, coldata_condition_col) %in%</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">23<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        colnames(coldata))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(ts_se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">45</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Prepare canonical condition order and base extraction.  `conds` is sorted</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # for deterministic ordering when constructing paired column order.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # `coldata_base` removes a trailing role suffix (e.g.  `_N`/`_T`) from the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # `coldata` sample IDs; note that assay column names may contain per-sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # annotations (e.g. per-q tags) which are handled separately below when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # building `sample_base_names`.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    conds &lt;- sort(unique(as.character(coldata[[coldata_condition_col]])))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    coldata_base &lt;- sub("_[^_]+$", "", as.character(coldata[[coldata_sample_col]]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    bases &lt;- unique(coldata_base)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Optionally validate that coldata defines pairing between conditions</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (isTRUE(paired) &amp;&amp; length(conds) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        unpaired &lt;- vapply(bases, function(b) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            length(unique(coldata[[coldata_condition_col]][coldata_base == b]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }, integer(1))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        bad &lt;- bases[unpaired != length(conds)]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(bad) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            bad_list &lt;- paste(bad, collapse = ", ")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            cond_list &lt;- paste(conds, collapse = ", ")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            msg &lt;- paste0("Unpaired samples found in coldata for bases: ", bad_list,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                ". Ensure each base has all conditions: ", cond_list)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop(msg, call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">68</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">69</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_base_names &lt;- sub("_q=.*", "", colnames(SummarizedExperiment::assay(ts_se)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">72</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Reorder the SummarizedExperiment columns.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    base_names &lt;- sample_base_names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx_list &lt;- integer(0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (isTRUE(paired)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Build paired-aware order: for each base, list samples in the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # deterministic `conds` order so matched samples become adjacent.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        for (b in bases) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            for (c in conds) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">80</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # find sample names in coldata for this base+condition</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                samples_for_pair &lt;- as.character(coldata[[coldata_sample_col]])[coldata_base ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">82</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  b &amp; as.character(coldata[[coldata_condition_col]]) == c]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (length(samples_for_pair) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">84</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  next</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">85</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">86</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # for each sample name, find matching assay columns (after</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">87</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # q-stripping)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                for (s in samples_for_pair) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  matches &lt;- which(base_names == s)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  if (length(matches) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    idx_list &lt;- c(idx_list, matches)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">92</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">93</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">95</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">96</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Preserve original order for any unmatched columns</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        remaining &lt;- setdiff(seq_along(base_names), idx_list)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">98</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        new_order &lt;- c(idx_list, remaining)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">99</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Non-paired: follow the order present in `coldata`</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">101</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ordered_samples &lt;- as.character(coldata[[coldata_sample_col]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        for (s in ordered_samples) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">91<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            matches &lt;- which(base_names == s)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">91<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(matches) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">79<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                idx_list &lt;- c(idx_list, matches)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">107</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">108</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        remaining &lt;- setdiff(seq_along(base_names), idx_list)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">109</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        new_order &lt;- c(idx_list, remaining)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">110</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(new_order) &gt; 0 &amp;&amp; !all(new_order == seq_along(base_names))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">112</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ts_se &lt;- ts_se[, new_order, drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">113</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_base_names &lt;- sub("_q=.*", "", colnames(SummarizedExperiment::assay(ts_se)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">114</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">115</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    st_map &lt;- setNames(as.character(coldata[[coldata_condition_col]]), as.character(coldata[[coldata_sample_col]]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">116</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_types &lt;- unname(st_map[sample_base_names])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">117</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    missing_idx &lt;- which(is.na(sample_types))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">118</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(missing_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">119</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        missing_samples &lt;- sample_base_names[missing_idx]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">120</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        msg &lt;- paste0("map_metadata: unmatched samples in 'coldata': ", paste(missing_samples,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">121</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            collapse = ", "), ". Provide matching entries in 'coldata' or populate ",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">122</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "colData(ts_se)$sample_type beforehand.")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">123</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(msg, call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">124</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    SummarizedExperiment::colData(ts_se)$sample_type &lt;- sample_types</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">126</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Also record the base sample names (without q suffix) per column so</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">127</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # downstream functions can rely on explicit pairing information.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">128</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    SummarizedExperiment::colData(ts_se)$sample_base &lt;- sample_base_names</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">129</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Ensure rownames of colData match assay column names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">130</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assay_cols &lt;- colnames(SummarizedExperiment::assay(ts_se))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">131</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rownames(SummarizedExperiment::colData(ts_se)) &lt;- assay_cols</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">132</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Attach transcript-level readcounts and tx-&gt;gene mapping to metadata if</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">133</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # they are available in the calling environment or globalenv and not</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">134</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # already present in the SummarizedExperiment metadata. This simplifies</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">135</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # downstream plotting helpers that expect these objects.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">136</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("S4Vectors", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">137</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md &lt;- S4Vectors::metadata(ts_se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # prefer existing metadata values; otherwise try common names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">139</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(md$readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">140</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (exists("readcounts", envir = parent.frame())) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">141</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$readcounts &lt;- get("readcounts", envir = parent.frame())</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">142</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else if (exists("readcounts", envir = globalenv())) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">143</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$readcounts &lt;- get("readcounts", envir = globalenv())</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">144</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">145</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(md$tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">147</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # vignette uses 'txmap' variable name; also accept 'tx2gene'</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">148</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (exists("txmap", envir = parent.frame())) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">149</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$tx2gene &lt;- get("txmap", envir = parent.frame())</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else if (exists("tx2gene", envir = parent.frame())) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">151</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$tx2gene &lt;- get("tx2gene", envir = parent.frame())</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">152</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else if (exists("txmap", envir = globalenv())) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">153</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$tx2gene &lt;- get("txmap", envir = globalenv())</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">154</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else if (exists("tx2gene", envir = globalenv())) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">155</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                md$tx2gene &lt;- get("tx2gene", envir = globalenv())</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">156</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">157</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">158</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        S4Vectors::metadata(ts_se) &lt;- md</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">159</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If a diversity assay is present, prepare a simple diversity data.frame</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">161</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # (genes + per-sample diversity values) and store it in metadata so the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">162</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # vignette and plotting helpers can use a ready-made table.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">163</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if ("diversity" %in% SummarizedExperiment::assayNames(ts_se)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">164</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        div_mat &lt;- as.matrix(SummarizedExperiment::assay(ts_se, "diversity"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">165</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # sample base names without per-q suffixes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">166</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_base_names &lt;- sub("_q=.*", "", colnames(div_mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">167</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # prefer explicit sample_type in colData when present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">168</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        samples_vec &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">169</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if ("sample_type" %in% colnames(SummarizedExperiment::colData(ts_se))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">170</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            samples_vec &lt;- as.character(SummarizedExperiment::colData(ts_se)$sample_type)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">171</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">172</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        div_df &lt;- as.data.frame(div_mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">173</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        genes_col &lt;- if (!is.null(SummarizedExperiment::rowData(ts_se)$genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">174</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            SummarizedExperiment::rowData(ts_se)$genes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">175</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">176</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            rownames(div_df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">177</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">178</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        div_df &lt;- cbind(genes = genes_col, div_df)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">179</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md2 &lt;- S4Vectors::metadata(ts_se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">180</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md2$diversity_df &lt;- div_df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">181</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md2$sample_base_names &lt;- sample_base_names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">182</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(samples_vec)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">183</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            md2$samples &lt;- samples_vec</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">184</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">185</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        S4Vectors::metadata(ts_se) &lt;- md2</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">186</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">187</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(ts_se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">188</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">189</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">190</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Map sample names (without '_q=...') to group labels using `colData(se)`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">191</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Mapping must be provided via `colData(se)`; no inference fallback is used.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">192</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">map_samples_to_group &lt;- function(sample_names, se = NULL, sample_type_col = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">193</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">194</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Prefer explicit mapping from colData(se)[, sample_type_col] when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">195</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # provided. If `sample_type_col` is not provided, allow a single- condition</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">196</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # dataset by assigning a single default group 'Group' to all samples (this</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">197</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # permits plotting single-condition q-curves).</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">198</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    base_names &lt;- sub("_q=.*", "", colnames(if (!is.null(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">199</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mat</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">200</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">201</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        SummarizedExperiment::assay(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">202</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">203</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">204</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(se) &amp;&amp; !is.null(sample_type_col) &amp;&amp; (sample_type_col %in% colnames(SummarizedExperiment::colData(se)))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">205</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_vec &lt;- as.character(SummarizedExperiment::colData(se)[, sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">206</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        names(st_vec) &lt;- base_names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">207</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_map &lt;- st_vec[!duplicated(names(st_vec))]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">208</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">209</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # No explicit mapping: assume single-group dataset</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">210</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_map &lt;- setNames(rep("Group", length(base_names)), base_names)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">211</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">212</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">213</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mapped &lt;- unname(st_map[sample_names])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">214</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    missing_idx &lt;- which(is.na(mapped))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">215</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(missing_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">216</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(sprintf("Missing sample_type mapping for samples: %s", paste(unique(sample_names[missing_idx]),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">217</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            collapse = ", ")))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">218</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">219</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mapped</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">220</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">221</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">222</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Prepare a long-format data.frame for a simple assay (one value per sample)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">223</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">get_assay_long &lt;- function(se, assay_name = "diversity", value_name = "diversity",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">224</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_type_col = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">225</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("tidyr", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">226</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("tidyr required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">227</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">228</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("dplyr", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">229</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("dplyr required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">230</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">231</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">232</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("SummarizedExperiment required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">233</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">234</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">235</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat &lt;- SummarizedExperiment::assay(se, assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">236</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">237</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Assay not found: ", assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">238</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">239</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- as.data.frame(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">240</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    genes_col &lt;- if (!is.null(SummarizedExperiment::rowData(se)$genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">241</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        SummarizedExperiment::rowData(se)$genes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">242</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">243</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">244</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">245</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- cbind(df, Gene = genes_col)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">246</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- tidyr::pivot_longer(df, -Gene, names_to = "sample", values_to = value_name)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">247</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">248</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # sample_type: prefer explicit colData mapping when available. If not</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">249</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # provided, assume a single-group dataset and set `sample_type` to 'Group'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">250</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # for all samples.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">251</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(sample_type_col) &amp;&amp; (sample_type_col %in% colnames(SummarizedExperiment::colData(se)))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">252</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st &lt;- as.character(SummarizedExperiment::colData(se)[, sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">253</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        names(st) &lt;- colnames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">254</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_map &lt;- st[!duplicated(names(st))]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">255</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_base &lt;- sub("_q=.*", "", long$sample)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">256</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$sample_type &lt;- unname(st_map[sample_base])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">257</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        missing_idx &lt;- which(is.na(long$sample_type))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">258</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(missing_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">259</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop(sprintf("Missing sample_type mapping for samples: %s", paste(unique(sample_base[missing_idx]),</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">260</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                collapse = ", ")))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">261</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">262</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">263</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$sample_type &lt;- rep("Group", nrow(long))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">264</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">265</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">266</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long[!is.na(long[[value_name]]), , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">267</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">268</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">269</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Internal small helper: prepare long-format tsallis data from a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">270</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># SummarizedExperiment</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">271</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">prepare_tsallis_long &lt;- function(se, assay_name = "diversity", sample_type_col = "sample_type") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">272</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("tidyr", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">273</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("tidyr required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">274</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">275</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("dplyr", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">276</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("dplyr required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">277</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">278</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">279</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("SummarizedExperiment required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">280</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">281</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">282</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat &lt;- SummarizedExperiment::assay(se, assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">283</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">284</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Assay not found: ", assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">285</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">286</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- as.data.frame(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">287</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    genes_col &lt;- if (!is.null(SummarizedExperiment::rowData(se)$genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">288</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        SummarizedExperiment::rowData(se)$genes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">289</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">290</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">291</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">292</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- cbind(df, Gene = genes_col)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">293</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">294</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    long &lt;- tidyr::pivot_longer(df, -Gene, names_to = "sample_q", values_to = "tsallis")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">295</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (any(grepl("_q=", long$sample_q))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">296</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long &lt;- tidyr::separate(long, sample_q, into = c("sample", "q"), sep = "_q=")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">297</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$q &lt;- as.factor(as.numeric(long$q))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">298</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">299</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$sample &lt;- long$sample_q</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">300</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$q &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">301</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">302</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">303</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(sample_type_col) &amp;&amp; (sample_type_col %in% colnames(SummarizedExperiment::colData(se)))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">304</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_vec &lt;- as.character(SummarizedExperiment::colData(se)[, sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">305</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        names(st_vec) &lt;- sub("_q=.*", "", colnames(mat))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">306</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st_map &lt;- st_vec[!duplicated(names(st_vec))]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">307</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$group &lt;- unname(st_map[as.character(long$sample)])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">308</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        missing_idx &lt;- which(is.na(long$group))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">309</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(missing_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">310</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop(sprintf("Missing sample_type mapping for samples: %s", paste(unique(as.character(long$sample)[missing_idx]),</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">311</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                collapse = ", ")))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">312</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">313</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">314</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        long$group &lt;- rep("Group", nrow(long))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">315</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">316</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">317</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    as.data.frame(long[!is.na(long$tsallis), , drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">318</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">319</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">320</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">321</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Map transcript IDs from a tx2gene table to a readcounts matrix</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">322</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">323</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Assign transcript identifiers as row names of a transcript-level read</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">324</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' counts matrix so downstream plotting functions can identify transcripts.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">325</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' The tx2gene mapping can be provided as a file path (TSV) or a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">326</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data.frame with a `Transcript` column.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">327</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">328</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param readcounts A numeric matrix or data.frame of read counts (rows = transcripts).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">329</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param tx2gene Either a path to a tab-delimited file or a data.frame with at least a `Transcript` column.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">330</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param tx_col Name of the transcript ID column in `tx2gene` (default: 'Transcript').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">331</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose Logical; print informative messages (default: FALSE).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">332</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return The input `readcounts` with rownames set to the transcript IDs.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">333</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">334</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' readcounts &lt;- matrix(1:4, nrow = 2, dimnames = list(NULL, c('s1', 's2')))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">335</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' tx2gene &lt;- data.frame(Transcript = c('tx1', 'tx2'), Gene = c('g1', 'g2'), stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">336</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- map_tx_to_readcounts(readcounts, tx2gene)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">337</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rownames(rc)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">338</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">339</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">map_tx_to_readcounts &lt;- function(readcounts, tx2gene, tx_col = "Transcript", verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">340</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(tx2gene) &amp;&amp; length(tx2gene) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">341</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!file.exists(tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">342</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("tx2gene file not found: ", tx2gene, call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">343</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">344</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txmap &lt;- utils::read.delim(tx2gene, header = TRUE, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">345</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.data.frame(tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">346</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txmap &lt;- tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">347</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">348</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("'tx2gene' must be a file path or a data.frame.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">349</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">350</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">351</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(tx_col %in% colnames(txmap))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">352</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(sprintf("tx2gene mapping must contain column '%s'", tx_col), call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">353</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">354</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">355</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Ensure readcounts is a matrix-like object</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">356</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.data.frame(readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">357</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        readcounts &lt;- as.matrix(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">358</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">359</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.matrix(readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">360</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("'readcounts' must be a matrix or data.frame", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">361</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">362</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">363</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    n_rc &lt;- nrow(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">364</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    n_tx &lt;- nrow(txmap)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">365</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">366</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (n_tx == n_rc) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">367</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(readcounts) &lt;- as.character(txmap[[tx_col]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">368</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">369</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            message(sprintf("Assigned %d transcript rownames from tx2gene.", n_rc))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">370</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">371</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">372</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">373</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">374</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If counts differ, attempt to match by transcript identifiers if present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">375</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tx_ids &lt;- as.character(txmap[[tx_col]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">376</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(rownames(readcounts)) &amp;&amp; all(rownames(readcounts) %in% tx_ids)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">377</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # reorder txmap to match readcounts row order and set rownames</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">378</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        matched_idx &lt;- match(rownames(readcounts), tx_ids)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">379</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(readcounts) &lt;- tx_ids[matched_idx]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">380</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">381</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            message("Matched and assigned transcript IDs by existing readcounts rownames.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">382</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">383</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">384</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">385</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">386</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    stop(sprintf("Number of transcripts in tx2gene (%d) does not match readcounts rows (%d), and automatic matching failed.",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">387</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        n_tx, n_rc), call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">388</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/calc_lm_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Summary reporting helper</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_report_fit_summary &lt;- function(res, verbose = TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">3</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (verbose &amp;&amp; "fit_method" %in% colnames(res)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">4</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        total_genes &lt;- nrow(res)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fallback_mask &lt;- !is.na(res$fit_method) &amp; res$fit_method != "lmer"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">6</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        n_fallback &lt;- sum(fallback_mask)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (n_fallback &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            tab &lt;- table(res$fit_method[fallback_mask])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">9</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            tab_str &lt;- paste(sprintf("%s=%d", names(tab), as.integer(tab)), collapse = ", ")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            message(sprintf("[calculate_lm_interaction] fallback fits used: %d/%d genes (%s)",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">11</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                n_fallback, total_genes, tab_str))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">13</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if ("singular" %in% colnames(res)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">14</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            n_sing &lt;- sum(as.logical(res$singular), na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">15</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (n_sing &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">16</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                message(sprintf("[calculate_lm_interaction] singular fits detected: %d/%d genes",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">17</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  n_sing, total_genes))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># GAM interaction helper</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_gam_interaction &lt;- function(df, q_vals, g, min_obs = 10) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">24</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("mgcv", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">25</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Package 'mgcv' is required for method = 'gam'")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uq_len &lt;- length(unique(na.omit(q_vals)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    k_q &lt;- max(2, min(10, uq_len - 1))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit_null &lt;- try(mgcv::gam(entropy ~ group + s(q, k = k_q), data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit_alt &lt;- try(mgcv::gam(entropy ~ group + s(q, by = group, k = k_q), data = df),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(fit_null, "try-error") || inherits(fit_alt, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">35</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    an &lt;- try(mgcv::anova.gam(fit_null, fit_alt, test = "F"), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(an, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">37</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p_interaction &lt;- NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(an) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if ("Pr(F)" %in% colnames(an)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">42</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_interaction &lt;- an[2, "Pr(F)"]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if ("Pr(&gt;F)" %in% colnames(an)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_interaction &lt;- an[2, "Pr(&gt;F)"]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if ("p-value" %in% colnames(an)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">46</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_interaction &lt;- an[2, "p-value"]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(data.frame(gene = g, p_interaction = p_interaction, stringsAsFactors = FALSE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># FPCA interaction helper</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_fpca_interaction &lt;- function(mat, q_vals, sample_names, group_vec, g, min_obs = 10) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uq &lt;- sort(unique(q_vals))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    samples_u &lt;- unique(sample_names)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    curve_mat &lt;- matrix(NA_real_, nrow = length(samples_u), ncol = length(uq))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(samples_u) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(curve_mat) &lt;- samples_u</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">59</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (i in seq_along(sample_names)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">88<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        s &lt;- sample_names[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">88<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        qv &lt;- q_vals[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">88<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        qi &lt;- match(qv, uq)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">88<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.na(qi)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            next</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">66</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">86<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (s %in% rownames(curve_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">86<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            curve_mat[s, qi] &lt;- as.numeric(mat[g, i])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">69</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">70</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    good_rows &lt;- which(rowSums(!is.na(curve_mat)) &gt;= max(2, ceiling(ncol(curve_mat)/2)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(good_rows) &lt; min_obs) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">74</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat_sub &lt;- curve_mat[good_rows, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    col_means &lt;- apply(mat_sub, 2, function(col) mean(col, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (r in seq_len(nrow(mat_sub))) mat_sub[r, is.na(mat_sub[r, ])] &lt;- col_means[is.na(mat_sub[r,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">78</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ])]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pca &lt;- try(stats::prcomp(mat_sub, center = TRUE, scale. = FALSE), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(pca, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">81</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (ncol(pca$x) &lt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">84</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">85</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pc1 &lt;- pca$x[, 1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    used_samples &lt;- rownames(mat_sub)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grp_vals &lt;- group_vec[match(used_samples, sample_names)]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(unique(na.omit(grp_vals))) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">90</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">91</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    g1 &lt;- unique(na.omit(grp_vals))[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    g2 &lt;- unique(na.omit(grp_vals))[2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">94</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x1 &lt;- pc1[grp_vals == g1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x2 &lt;- pc1[grp_vals == g2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(x1) &lt; 2 || length(x2) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">97</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    t_res &lt;- try(stats::t.test(x1, x2), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(t_res, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">101</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">102</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pval &lt;- as.numeric(t_res$p.value)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(data.frame(gene = g, p_interaction = pval, stringsAsFactors = FALSE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">105</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">107</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Fit function extracted from calculate_lm_interaction</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">108</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_fit_one_interaction &lt;- function(g, se, mat, q_vals, sample_names, group_vec,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">109</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    method, pvalue, subject_col, paired, min_obs, verbose, suppress_lme4_warnings,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">110</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    progress) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">77<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    vals &lt;- as.numeric(mat[g, ])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">112</td>
                    <td class="coverage">77<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- data.frame(entropy = vals, q = q_vals, group = factor(group_vec))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">113</td>
                    <td class="coverage">77<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (sum(!is.na(df$entropy)) &lt; min_obs) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">114</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">115</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">116</td>
                    <td class="coverage">75<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(unique(na.omit(df$group))) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">117</td>
                    <td class="coverage">51<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">118</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">119</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">120</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "linear") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">121</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit &lt;- try(stats::lm(entropy ~ q * group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">122</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (inherits(fit, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">123</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">124</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        coefs &lt;- summary(fit)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">126</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">127</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(ia_idx) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">128</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">129</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">130</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p_interaction &lt;- coefs[ia_idx[1], "Pr(&gt;|t|)"]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">131</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(data.frame(gene = g, p_interaction = p_interaction, stringsAsFactors = FALSE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">132</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">133</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">134</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "lmm") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">135</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!requireNamespace("lme4", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">136</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Package 'lme4' is required for method = 'lmm'")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">137</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # determine subject IDs for random effect</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">139</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        subject &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">140</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(subject_col)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">141</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!(subject_col %in% colnames(SummarizedExperiment::colData(se)))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">142</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop(sprintf("subject_col '%s' not found in colData(se)", subject_col))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">143</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">144</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            subj_full &lt;- as.character(SummarizedExperiment::colData(se)[, subject_col])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">145</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            names(subj_full) &lt;- SummarizedExperiment::colData(se)$samples %||% colnames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            subject &lt;- unname(subj_full[sample_names])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">147</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if (paired) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">148</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # require a paired identifier (sample_base) in colData</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">149</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if ("sample_base" %in% colnames(SummarizedExperiment::colData(se))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                sb &lt;- as.character(SummarizedExperiment::colData(se)[, "sample_base"])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">151</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                names(sb) &lt;- SummarizedExperiment::colData(se)$samples %||% colnames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">152</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                subject &lt;- unname(sb[sample_names])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">153</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">154</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("paired = TRUE but no 'sample_base' column found in colData(se); call map_metadata(..., paired = TRUE) or supply subject_col")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">155</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">156</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">157</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # fallback to sample base derived from column names</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">158</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            subject &lt;- sample_names</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">159</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">160</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df$subject &lt;- factor(subject)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">161</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # require at least two subjects and at least two groups represented</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">162</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(unique(na.omit(df$subject))) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">163</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">164</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">165</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(unique(na.omit(df$group))) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">166</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">167</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">168</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # fit null (no interaction) and alternative (with q:group interaction)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">169</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mm_suppress_pattern &lt;- "boundary \\(singular\\) fit|Computed variance-covariance matrix problem|not a positive definite matrix"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">170</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">171</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # attempt to fit mixed models (null and alternative) using helper</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">172</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit0 &lt;- .tsenat_try_lmer(entropy ~ q + group + (1 | subject), df, suppress_lme4_warnings = suppress_lme4_warnings,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">173</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            verbose = verbose, mm_suppress_pattern = mm_suppress_pattern)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">174</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit1 &lt;- .tsenat_try_lmer(entropy ~ q * group + (1 | subject), df, suppress_lme4_warnings = suppress_lme4_warnings,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">175</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            verbose = verbose, mm_suppress_pattern = mm_suppress_pattern)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">176</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">177</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # If either fit failed, try falling back to simpler approaches</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">178</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fallback_lm &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">179</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        used_fit_method &lt;- "lmer"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">180</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        used_singular &lt;- FALSE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">181</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (inherits(fit0, "try-error") || inherits(fit1, "try-error") || (inherits(fit0,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">182</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "lmerMod") &amp;&amp; attr(fit0, "singular") == TRUE) || (inherits(fit1, "lmerMod") &amp;&amp;</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">183</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            attr(fit1, "singular") == TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">184</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            used_fit_method &lt;- "fallback"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">185</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            used_singular &lt;- (inherits(fit0, "lmerMod") &amp;&amp; attr(fit0, "singular") ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">186</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                TRUE) || (inherits(fit1, "lmerMod") &amp;&amp; attr(fit1, "singular") ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">187</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">188</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if ((verbose &amp;&amp; progress) || (!verbose &amp;&amp; progress)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">189</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                message("[calculate_lm_interaction] mixed model singular or failed; trying simpler fixed-effects fallback")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">190</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">191</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            fb &lt;- .tsenat_try_lm_fallbacks(df, verbose = verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">192</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.null(fb)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">193</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                fallback_lm &lt;- fb</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">194</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                used_fit_method &lt;- fb$method</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">195</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">196</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">197</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            used_singular &lt;- (inherits(fit0, "lmerMod") &amp;&amp; attr(fit0, "singular") ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">198</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                TRUE) || (inherits(fit1, "lmerMod") &amp;&amp; attr(fit1, "singular") ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">199</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">200</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (used_singular) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">201</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                used_fit_method &lt;- "lmer_singular"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">202</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">203</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                used_fit_method &lt;- "lmer"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">204</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">205</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">206</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">207</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        lrt_p &lt;- NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">208</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(fallback_lm)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">209</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            lrt_p &lt;- .tsenat_extract_lrt_p(fallback_lm$fit0, fallback_lm$fit1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">210</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">211</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            lrt_p &lt;- .tsenat_extract_lrt_p(fit0, fit1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">212</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">213</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">214</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        satter_p &lt;- NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">215</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (pvalue %in% c("satterthwaite", "both")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">216</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            mm_suppress_pattern &lt;- "boundary \\(singular\\) fit|Computed variance-covariance matrix problem|not a positive definite matrix"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">217</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            muffle_cond &lt;- suppress_lme4_warnings || (!verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">218</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            satter_p &lt;- withCallingHandlers(.tsenat_extract_satterthwaite_p(fit1,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">219</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                fallback_lm), warning = function(w) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">220</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(w),</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">221</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">222</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  invokeRestart("muffleWarning")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">223</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">224</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }, message = function(m) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">225</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(m),</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">226</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">227</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  invokeRestart("muffleMessage")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">228</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">229</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            })</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">230</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">231</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">232</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # choose primary p_interaction according to user preference</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">233</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p_interaction &lt;- NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">234</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (pvalue == "satterthwaite") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">235</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_interaction &lt;- satter_p</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">236</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if (pvalue == "lrt") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">237</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_interaction &lt;- lrt_p</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">238</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if (pvalue == "both") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">239</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.na(satter_p)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">240</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                p_interaction &lt;- satter_p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">241</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">242</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                p_interaction &lt;- lrt_p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">243</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">244</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">245</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">246</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(data.frame(gene = g, p_interaction = p_interaction, p_lrt = lrt_p,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">247</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_satterthwaite = satter_p, fit_method = used_fit_method, singular = used_singular,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">248</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stringsAsFactors = FALSE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">249</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">250</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">251</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "gam") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">252</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(.tsenat_gam_interaction(df, q_vals, g, min_obs = min_obs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">253</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">254</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">255</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "fpca") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">256</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(.tsenat_fpca_interaction(mat, q_vals, sample_names, group_vec, g,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">257</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            min_obs = min_obs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">258</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">259</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">260</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">261</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## All helpers for calculate_lm_interaction</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">262</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">263</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Try lme4::lmer with multiple optimizers and controlled warnings.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">264</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_try_lmer &lt;- function(formula, data, suppress_lme4_warnings = TRUE, verbose = FALSE,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">265</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mm_suppress_pattern = "boundary \\(singular\\) fit|Computed variance-covariance matrix problem|not a positive definite matrix") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">266</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("lme4", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">267</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Package 'lme4' is required for mixed-model fitting")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">268</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">269</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    opts &lt;- list(list(optimizer = "bobyqa", optCtrl = list(maxfun = 2e+05)), list(optimizer = "nloptwrap",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">270</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        optCtrl = list(maxfun = 5e+05)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">271</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (o in opts) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">272</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ctrl &lt;- lme4::lmerControl(optimizer = o$optimizer, optCtrl = o$optCtrl)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">273</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        muffle_cond &lt;- suppress_lme4_warnings || (!verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">274</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit_try &lt;- withCallingHandlers(try(lme4::lmer(formula, data = data, REML = FALSE,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">275</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            control = ctrl), silent = TRUE), warning = function(w) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">276</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(w), ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">277</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                invokeRestart("muffleWarning")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">278</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">279</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }, message = function(m) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">280</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(m), ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">281</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                invokeRestart("muffleMessage")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">282</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">283</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        })</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">284</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(fit_try, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">285</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # check singularity if function available</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">286</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            is_sing &lt;- FALSE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">287</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (exists("isSingular", where = asNamespace("lme4"), inherits = FALSE)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">288</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                is_sing &lt;- tryCatch(lme4::isSingular(fit_try, tol = 1e-04), error = function(e) FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">289</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">290</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            attr(fit_try, "singular") &lt;- is_sing</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">291</td>
                    <td class="coverage">17<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(fit_try)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">292</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">293</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">294</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # all attempts failed</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">295</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(structure("error", class = "try-error"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">296</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">297</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">298</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># LM fallback helpers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">299</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_try_lm_fallbacks &lt;- function(df, verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">300</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm &lt;- try(stats::lm(entropy ~ q + group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">301</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm &lt;- try(stats::lm(entropy ~ q * group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">302</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm, "try-error") &amp;&amp; !inherits(fit1_lm, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">303</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm, fit1 = fit1_lm, method = "lm_subject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">304</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">305</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm2 &lt;- try(stats::lm(entropy ~ q + group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">306</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm2 &lt;- try(stats::lm(entropy ~ q * group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">307</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm2, "try-error") &amp;&amp; !inherits(fit1_lm2, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">308</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm2, fit1 = fit1_lm2, method = "lm_nosubject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">309</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">310</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">311</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">312</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">313</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># LRT p-value extraction</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">314</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_lrt_p &lt;- function(fit0, fit1) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">315</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    an &lt;- try(stats::anova(fit0, fit1), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">316</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(an, "try-error") &amp;&amp; nrow(an) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">317</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pcol &lt;- grep("Pr\\(&gt;F\\)|Pr\\(&gt;Chisq\\)|Pr\\(&gt;Chi\\)", colnames(an), value = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">318</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(pcol) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">319</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.numeric(an[2, ncol(an)]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">320</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">321</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.numeric(an[2, pcol[1]]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">322</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">323</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">324</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">325</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">326</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">327</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Satterthwaite p-value extraction</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">328</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_satterthwaite_p &lt;- function(fit1, fallback_lm = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">329</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(fallback_lm)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">330</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        coefs &lt;- try(summary(fallback_lm$fit1)$coefficients, silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">331</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(coefs) &amp;&amp; !inherits(coefs, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">332</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">333</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">334</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">335</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">336</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">337</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">338</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">339</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("lmerTest", quietly = TRUE) &amp;&amp; inherits(fit1, "lmerMod") &amp;&amp;</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">340</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        !isTRUE(attr(fit1, "singular"))) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">341</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit_lt &lt;- try(lmerTest::lmer(stats::formula(fit1), data = stats::model.frame(fit1),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">342</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            REML = FALSE), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">343</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(fit_lt, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">344</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            coefs &lt;- summary(fit_lt)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">345</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">346</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">347</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">348</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">349</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">350</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">351</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">352</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">353</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_satterthwaite_p &lt;- function(fit1, fallback_lm = NULL, suppress_lme4_warnings = TRUE,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">354</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    verbose = FALSE, mm_suppress_pattern = "boundary \\(singular\\) fit|Computed variance-covariance matrix problem|not a positive definite matrix") {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">355</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If we have a fallback lm, extract from its coefficients</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">356</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(fallback_lm)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">357</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        coefs &lt;- try(summary(fallback_lm$fit1)$coefficients, silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">358</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(coefs, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">359</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">360</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">361</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">362</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">363</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">364</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">365</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">366</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Prefer lmerTest when available; suppress known lme4/lmerTest warnings</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">367</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("lmerTest", quietly = TRUE) &amp;&amp; inherits(fit1, "lmerMod")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">368</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        muffle_cond &lt;- suppress_lme4_warnings || (!verbose)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">369</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit_lt &lt;- withCallingHandlers(try(lmerTest::lmer(stats::formula(fit1), data = stats::model.frame(fit1),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">370</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            REML = FALSE), silent = TRUE), warning = function(w) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">371</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(w), ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">372</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                invokeRestart("muffleWarning")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">373</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">374</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }, message = function(m) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">375</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (muffle_cond &amp;&amp; grepl(mm_suppress_pattern, conditionMessage(m), ignore.case = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">376</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                invokeRestart("muffleMessage")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">377</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">378</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        })</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">379</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(fit_lt, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">380</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            coefs &lt;- summary(fit_lt)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">381</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">382</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">383</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">384</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">385</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">386</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">387</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">388</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">389</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">390</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># FPCA matrix preparation</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">391</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_fpca_matrix &lt;- function(mat, min_frac = 0.01) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">392</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.matrix(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">393</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mat &lt;- as.matrix(mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">394</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">395</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    row_vars &lt;- apply(mat, 1, stats::var, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">396</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    keep &lt;- row_vars &gt; (min_frac * max(row_vars, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">397</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (sum(keep) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">398</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        keep &lt;- rep(TRUE, nrow(mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">399</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">400</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    m2 &lt;- mat[keep, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">401</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    m2 &lt;- t(scale(t(m2)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">402</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(list(mat = m2, keep = keep))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">403</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">404</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">405</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Helpers for calculate_lm_interaction fallbacks, LRT and Satterthwaite</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">406</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## p-values</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">407</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_try_lm_fallbacks &lt;- function(df, verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">408</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # try lm with subject as fixed effect</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">409</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm &lt;- try(stats::lm(entropy ~ q + group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">410</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm &lt;- try(stats::lm(entropy ~ q * group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">411</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm, "try-error") &amp;&amp; !inherits(fit1_lm, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">412</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm, fit1 = fit1_lm, method = "lm_subject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">413</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">414</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # last resort: drop subject, plain lm</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">415</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm2 &lt;- try(stats::lm(entropy ~ q + group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">416</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm2 &lt;- try(stats::lm(entropy ~ q * group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">417</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm2, "try-error") &amp;&amp; !inherits(fit1_lm2, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">418</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm2, fit1 = fit1_lm2, method = "lm_nosubject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">419</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">420</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">421</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">422</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">423</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_lrt_p &lt;- function(fit0, fit1) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">424</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    an &lt;- try(stats::anova(fit0, fit1), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">425</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(an, "try-error") &amp;&amp; nrow(an) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">426</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pcol &lt;- grep("Pr\\(&gt;F\\)|Pr\\(&gt;Chisq\\)|Pr\\(&gt;Chi\\)", colnames(an), value = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">427</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(pcol) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">428</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.numeric(an[2, ncol(an)]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">429</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">430</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(as.numeric(an[2, pcol[1]]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">431</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">432</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">433</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">434</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">435</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">436</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_satterthwaite_p &lt;- function(fit1, fallback_lm = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">437</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # if lmerTest available and fit1 is not singular use that</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">438</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(fallback_lm)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">439</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        coefs &lt;- try(summary(fallback_lm$fit1)$coefficients, silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">440</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(coefs, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">441</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">442</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">443</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">444</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">445</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">446</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">447</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">448</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("lmerTest", quietly = TRUE) &amp;&amp; inherits(fit1, "lmerMod") &amp;&amp;</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">449</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        !isTRUE(attr(fit1, "singular"))) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">450</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit_lt &lt;- try(lmerTest::lmer(stats::formula(fit1), data = stats::model.frame(fit1),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">451</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            REML = FALSE), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">452</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(fit_lt, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">453</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            coefs &lt;- summary(fit_lt)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">454</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">455</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">456</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">457</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">458</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">459</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">460</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">461</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">462</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">463</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_fpca_matrix &lt;- function(mat, min_frac = 0.01) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">464</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # prepare matrix for FPCA: center, scale, and drop near-constant rows</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">465</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.matrix(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">466</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mat &lt;- as.matrix(mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">467</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">468</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    row_vars &lt;- apply(mat, 1, stats::var, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">469</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    keep &lt;- row_vars &gt; (min_frac * max(row_vars, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">470</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (sum(keep) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">471</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        keep &lt;- rep(TRUE, nrow(mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">472</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">473</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    m2 &lt;- mat[keep, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">474</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    m2 &lt;- t(scale(t(m2)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">475</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(list(mat = m2, keep = keep))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">476</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">477</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Additional helpers for calculate_lm_interaction</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">478</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">479</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Try simple lm fallbacks when lmer fails. Returns list(fit0, fit1, method)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">480</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_try_lm_fallbacks &lt;- function(df, verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">481</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # try lm with subject as fixed effect</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">482</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm &lt;- try(stats::lm(entropy ~ q + group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">483</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm &lt;- try(stats::lm(entropy ~ q * group + subject, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">484</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm, "try-error") &amp;&amp; !inherits(fit1_lm, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">485</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm, fit1 = fit1_lm, method = "lm_subject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">486</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">487</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # last resort: drop subject</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">488</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit0_lm2 &lt;- try(stats::lm(entropy ~ q + group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">489</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit1_lm2 &lt;- try(stats::lm(entropy ~ q * group, data = df), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">490</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!inherits(fit0_lm2, "try-error") &amp;&amp; !inherits(fit1_lm2, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">491</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(list(fit0 = fit0_lm2, fit1 = fit1_lm2, method = "lm_nosubject"))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">492</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">493</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # indicate failure</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">494</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">495</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">496</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">497</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Extract LRT p-value from two fitted models (lm or lmer). Returns NA_real_ on</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">498</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># failure.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">499</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_lrt_p &lt;- function(fit0, fit1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">500</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    an &lt;- try(stats::anova(fit0, fit1), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">501</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (inherits(an, "try-error") || nrow(an) &lt; 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">502</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">503</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">504</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pcol &lt;- grep("Pr\\(&gt;F\\)|Pr\\(&gt;Chisq\\)|Pr\\(&gt;Chi\\)", colnames(an), value = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">505</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(pcol) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">506</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(as.numeric(an[2, ncol(an)]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">507</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">508</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(as.numeric(an[2, pcol[1]]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">509</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">510</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">511</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Extract Satterthwaite p-value for q:group interaction from lmer/lmerTest or</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">512</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># fallback lm.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">513</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_extract_satterthwaite_p &lt;- function(fit1, fallback_lm = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">514</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # prefer lmerTest when available</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">515</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(fallback_lm)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">516</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        coefs &lt;- summary(fallback_lm$fit1)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">517</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">518</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">519</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">520</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">521</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">522</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">523</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (requireNamespace("lmerTest", quietly = TRUE) &amp;&amp; inherits(fit1, "lmerMod")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">524</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        fit_lt &lt;- try(lmerTest::lmer(as.formula(formula(fit1)), data = model.frame(fit1),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">525</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            REML = FALSE), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">526</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!inherits(fit_lt, "try-error")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">527</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            coefs &lt;- summary(fit_lt)$coefficients</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">528</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ia_idx &lt;- grep("^q:group", rownames(coefs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">529</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (length(ia_idx) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">530</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                return(coefs[ia_idx[1], "Pr(&gt;|t|)"])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">531</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">532</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">533</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">534</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(NA_real_)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">535</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">536</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">537</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Helper for FPCA-style preprocessing used in calculate_lm_interaction fpca</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">538</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># method.  Builds curve_mat, filters good rows, imputes column means, and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">539</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># returns list(mat_sub, used_samples)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">540</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_fpca_matrix &lt;- function(mat, sample_names, q_vals, min_obs = 10) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">541</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uq &lt;- sort(unique(q_vals))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">542</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    samples_u &lt;- unique(sample_names)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">543</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    curve_mat &lt;- matrix(NA_real_, nrow = length(samples_u), ncol = length(uq))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">544</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(samples_u) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">545</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(curve_mat) &lt;- samples_u</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">546</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">547</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (i in seq_along(sample_names)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">548</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        s &lt;- sample_names[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">549</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        qv &lt;- q_vals[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">550</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        qi &lt;- match(qv, uq)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">551</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.na(qi)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">552</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            next</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">553</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">554</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (s %in% rownames(curve_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">555</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            curve_mat[s, qi] &lt;- as.numeric(mat[, i])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">556</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">557</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">558</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # keep samples with at least half of q points present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">559</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    good_rows &lt;- which(rowSums(!is.na(curve_mat)) &gt;= max(2, ceiling(ncol(curve_mat)/2)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">560</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(good_rows) &lt; min_obs) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">561</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">562</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">563</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat_sub &lt;- curve_mat[good_rows, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">564</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    col_means &lt;- apply(mat_sub, 2, function(col) mean(col, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">565</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (r in seq_len(nrow(mat_sub))) mat_sub[r, is.na(mat_sub[r, ])] &lt;- col_means[is.na(mat_sub[r,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">566</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ])]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">567</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(mat_sub = mat_sub, used_samples = rownames(mat_sub))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">568</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/plot_top_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Helpers for plot_top_transcripts internals</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_select_genes_from_res &lt;- function(res, top_n) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">4</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(res))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Either 'gene' or 'res' must be provided")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">6</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!("genes" %in% colnames(res)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Provided 'res' must contain a 'genes' column")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if ("adjusted_p_values" %in% colnames(res)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">9</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ord &lt;- order(res$adjusted_p_values, na.last = NA)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if ("raw_p_values" %in% colnames(res)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">11</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ord &lt;- order(res$raw_p_values, na.last = NA)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">13</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ord &lt;- seq_len(nrow(res))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">15</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    genes_sel &lt;- as.character(res$genes[ord])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">16</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    genes_sel &lt;- unique(genes_sel)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">17</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    head(genes_sel, top_n)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_infer_samples_from_coldata &lt;- function(coldata, counts, sample_type_col) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">21</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(coldata) &amp;&amp; length(coldata) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">22</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!file.exists(coldata))</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">23</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("coldata file not found: ", coldata)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">24</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        cdf &lt;- utils::read.delim(coldata, header = TRUE, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">25</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.data.frame(coldata)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">26</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        cdf &lt;- coldata</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`coldata` must be a data.frame or path to a tab-delimited file")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(rownames(cdf)) &amp;&amp; all(colnames(counts) %in% rownames(cdf))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        as.character(cdf[colnames(counts), sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_id_cols &lt;- c("sample", "Sample", "sample_id", "id")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">35</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sid &lt;- intersect(sample_id_cols, colnames(cdf))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(sid) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            sid &lt;- sid[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!all(colnames(counts) %in% as.character(cdf[[sid]])))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("coldata sample id column does not match column names of counts")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            row_ix &lt;- match(colnames(counts), as.character(cdf[[sid]]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            as.character(cdf[[sample_type_col]][row_ix])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Could not match `coldata` rows to `counts` columns. Provide `samples` or a row-named `coldata`.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">44</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">45</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_read_tx2gene &lt;- function(tx2gene) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(tx2gene))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`tx2gene` must be provided as a file path or data.frame (or include mapping in metadata of provided SummarizedExperiment)")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(tx2gene) &amp;&amp; length(tx2gene) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!file.exists(tx2gene))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("tx2gene file not found: ", tx2gene)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mapping &lt;- utils::read.delim(tx2gene, stringsAsFactors = FALSE, header = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.data.frame(tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mapping &lt;- tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">58</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`tx2gene` must be provided as a file path or data.frame (or include mapping in metadata of provided SummarizedExperiment)")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">59</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!all(c("Transcript", "Gen") %in% colnames(mapping)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("tx2gene must have columns 'Transcript' and 'Gen'")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mapping</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">65</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_make_agg &lt;- function(metric = c("median", "mean", "variance", "iqr")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    metric_choice &lt;- match.arg(metric)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_fun &lt;- switch(metric_choice, median = function(x) stats::median(x, na.rm = TRUE),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        mean = function(x) base::mean(x, na.rm = TRUE), variance = function(x) stats::var(x,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            na.rm = TRUE), iqr = function(x) stats::IQR(x, na.rm = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label_metric &lt;- if (metric_choice == "iqr")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "IQR" else metric_choice</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label &lt;- sprintf("Transcript-level expression with metric %s", agg_label_metric)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .cnt &lt;- as.integer(getOption("TSENAT.plot_top_counter", 0)) + 1L</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    options(TSENAT.plot_top_counter = .cnt)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg_label_unique &lt;- agg_label</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(metric_choice = metric_choice, agg_fun = agg_fun, agg_label_unique = agg_label_unique)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">78</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">79</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_build_tx_long &lt;- function(gene_single, mapping, counts, samples, top_n) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">19<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    txs &lt;- mapping$Transcript[mapping$Gen == gene_single]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">19<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    txs &lt;- intersect(txs, rownames(counts))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">82</td>
                    <td class="coverage">19<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(txs) == 0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("No transcripts found for gene: ", gene_single)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(top_n))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txs &lt;- head(txs, top_n)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat &lt;- counts[txs, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_all &lt;- as.data.frame(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_all$tx &lt;- rownames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_long &lt;- tidyr::pivot_longer(df_all, -tx, names_to = "sample", values_to = "expr")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_long$group &lt;- rep(samples, times = length(txs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(df_long = df_long, txs = txs)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">92</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">93</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_aggregate_df_long &lt;- function(df_long, agg_fun, pseudocount) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_summary &lt;- stats::aggregate(expr ~ tx + group, data = df_long, FUN = agg_fun)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_summary$log2expr &lt;- log2(df_summary$expr + pseudocount)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_summary$tx &lt;- factor(df_summary$tx, levels = unique(df_summary$tx))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">98</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_summary</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">99</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_build_plot_from_summary &lt;- function(df_summary, agg_label_unique, fill_limits = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p &lt;- ggplot2::ggplot(df_summary, ggplot2::aes(x = group, y = tx, fill = log2expr)) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::geom_tile(color = "white", width = 0.95, height = 0.95) + ggplot2::scale_fill_viridis_c(option = "viridis",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        direction = -1, na.value = "grey80", limits = fill_limits) + ggplot2::theme_minimal(base_size = 14) +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::labs(title = agg_label_unique, x = NULL, y = NULL, fill = "log2(expr)") +</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">106</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(axis.text.y = ggplot2::element_text(size = 12), axis.text.x = ggplot2::element_text(size = 12),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            plot.title = ggplot2::element_text(size = 16, hjust = 0.6), legend.position = "bottom",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">108</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            legend.key.width = ggplot2::unit(1.2, "cm"), plot.margin = ggplot2::margin(4,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">109</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                4, 4, 4)) + ggplot2::guides(fill = ggplot2::guide_colorbar(title.position = "top",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">110</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        barwidth = 6, barheight = 0.35))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">18<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">112</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">113</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">114</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_combine_patchwork &lt;- function(plots, agg_label_unique) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">115</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    combined &lt;- Reduce(`+`, plots) + patchwork::plot_layout(nrow = 1, guides = "collect") &amp;</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">116</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::theme(legend.position = "bottom")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">117</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    combined &lt;- combined + patchwork::plot_annotation(title = agg_label_unique, theme = ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.6,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">118</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        size = 16, margin = ggplot2::margin(b = 10))))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">119</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    combined</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">120</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">121</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">122</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_combine_cowplot &lt;- function(plots, output_file = NULL, agg_label_unique) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">123</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p_for_legend &lt;- plots[[1]] + ggplot2::theme(legend.position = "bottom")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">124</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    legend &lt;- cowplot::get_legend(p_for_legend)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plots_nolegend &lt;- lapply(plots, function(pp) pp + ggplot2::theme(legend.position = "none"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">126</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grid &lt;- cowplot::plot_grid(plotlist = plots_nolegend, nrow = 1, align = "hv")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">127</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    title_grob &lt;- cowplot::ggdraw() + cowplot::draw_label(agg_label_unique, fontface = "plain",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">128</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x = 0.6, hjust = 0.5, size = 16)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">129</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_plot &lt;- cowplot::plot_grid(title_grob, grid, legend, ncol = 1, rel_heights = c(0.08,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">130</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        1, 0.08))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">131</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(output_file)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">132</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ggplot2::ggsave(output_file, result_plot)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">133</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        invisible(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">134</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">135</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_plot</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">136</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">137</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.ptt_combine_grid &lt;- function(plots, output_file = NULL, agg_label_unique) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">139</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plots_nolegend &lt;- lapply(plots, function(pp) pp + ggplot2::theme(legend.position = "none"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">140</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    grobs &lt;- lapply(plots_nolegend, ggplot2::ggplotGrob)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">141</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    g_full &lt;- ggplot2::ggplotGrob(plots[[1]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">142</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    legend_idx &lt;- which(vapply(g_full$grobs, function(x) x$name, character(1)) ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">143</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "guide-box")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">144</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(legend_idx)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">145</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        legend_grob &lt;- g_full$grobs[[legend_idx[1]]]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">146</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">147</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        legend_grob &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">148</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">149</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    n &lt;- length(grobs)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    heights &lt;- grid::unit.c(grid::unit(0.6, "cm"), grid::unit(1, "null"), grid::unit(0.7,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">151</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "cm"))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">152</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(output_file)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">153</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        png(filename = output_file, width = 800 * n, height = 480, res = 150)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">154</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .draw_transcript_grid(grobs, agg_label_unique, legend_grob, n, heights, to_file = output_file)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">155</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        invisible(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">156</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">157</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .draw_transcript_grid(grobs, agg_label_unique, legend_grob, n, heights)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">158</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        invisible(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">159</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/difference_functions.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate splicing diversity changes between two conditions.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A \code{matrix} with the splicing diversity values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Character vector with an equal length to the number of columns</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' in the input dataset, specifying the category of each sample.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param control Name of the control sample category, defined in the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{samples} vector, e.g. \code{control = 'Normal'} or \code{control =</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' 'WT'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param method Method to use for calculating the average splicing diversity</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' value in a condition. Can be \code{'mean'} or \code{'median'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pseudocount Numeric scalar. Small value added to non-positive</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' observed group summaries to avoid zeros when computing differences and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' log2 fold-changes. If \code{pseudocount &lt;= 0} the function will automatically</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' choose a scale-aware value equal to half the smallest positive observed</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' group summary (i.e. half the smallest observed mean/median across groups);</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' if no positive values are present the fallback is \code{1e-6}. Rows with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' insufficient observations remain \code{NA} and are not imputed.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A \code{data.frame} with mean or median value of splicing diversity</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' across sample categories, the difference between these values and the log2</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' fold change values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @details The function uses a matrix of splicing diversity values in order to</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' calculate mean or median differences and log2 fold changes between two</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' conditions.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import stats</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_fc &lt;- function(x, samples, control, method = "mean", pseudocount = 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # validate control and samples inputs</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">553<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(control) || !nzchar(control)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`control` must be provided to calculate_fc", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">551<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(samples) != ncol(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Length of 'samples' must equal number of columns in 'x'", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">548<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(control %in% samples)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Control sample type not found in samples.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    agg &lt;- .tsenat_aggregate_fc_values(x = x, samples = samples, method = method,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        control = control)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value &lt;- agg$value</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sorted &lt;- agg$sorted</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">40</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Defensive numeric coercion: ensure group means are numeric and mark any</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # non-finite or non-positive values as NA. This prevents Inf/NaN when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">43</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # computing log2 fold changes downstream.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value &lt;- matrix(as.numeric(value), nrow = nrow(value), ncol = ncol(value), dimnames = dimnames(value))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value[!is.finite(value)] &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # compute and apply pseudocount based on observed group summaries</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">48</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value &lt;- .tsenat_apply_pseudocount(value, pseudocount)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # compute difference and log2 fold-change with NA-safe handling</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    diff_vec &lt;- value[, 1] - value[, 2]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    na_mask &lt;- is.na(value[, 1]) | is.na(value[, 2])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    diff_vec[na_mask] &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    log2fc_vec &lt;- log2(value[, 1]/value[, 2])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    log2fc_vec[na_mask] &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result &lt;- data.frame(value, difference = diff_vec, log2_fold_change = log2fc_vec,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        check.names = FALSE, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(result) &lt;- c(paste(sorted[1, 1], "_", method, sep = ""), paste(sorted[2,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        1], "_", method, sep = ""), paste(method, "_difference", sep = ""), "log2_fold_change")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">546<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(result)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">65</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate p-values using Wilcoxon rank sum test.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">66</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">67</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A \code{matrix} with the splicing diversity values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">68</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Character vector with an equal length to the number of columns</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">69</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' in the input dataset, specifying the category of each sample.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">70</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pcorr P-value correction method applied to the results, as defined in</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the \code{p.adjust} function.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">72</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired If \code{TRUE}, the Wilcox-test will be paired, and therefore</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">73</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' it will be a signed rank test instead of the rank sum test.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">74</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param exact If \code{TRUE}, an exact p-value will be computed.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">75</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return Raw and corrected p-values in a matrix.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import stats</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">wilcoxon &lt;- function(x, samples, pcorr = "BH", paired = FALSE, exact = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">78</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Determine group indices (two groups expected)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    groups &lt;- unique(sort(samples))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(groups) != 2) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">81</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`samples` must contain exactly two groups for Wilcoxon tests.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    g1_idx &lt;- as.numeric(which(samples %in% groups[1]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    g2_idx &lt;- as.numeric(which(samples %in% groups[2]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">85</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (isTRUE(paired)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(g1_idx) != length(g2_idx)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("Paired Wilcoxon requires equal numbers of samples in each ", "group.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">89</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">90</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Paired tests assume columns are already ordered/aligned by the caller</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">91</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # (e.g., via `map_metadata()`); do not attempt to infer pairing here.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">92</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">93</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">94</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p_values &lt;- vector("list", nrow(x))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    for (i in seq_len(nrow(x))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        p_values[i] &lt;- tryCatch({</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            wilcox.test(x[i, g1_idx], x[i, g2_idx], paired = paired, exact = exact)$p.value</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">98</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }, error = function(e) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }, warning = function(w) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # swallow specific warnings but return NA on unusual states</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">55<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            NA_real_</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">103</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        })</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">104</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">105</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">106</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    raw_p_values &lt;- ifelse(is.na(vapply(p_values, c, numeric(1))), 1, vapply(p_values,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        c, numeric(1)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">108</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    adjusted_p_values &lt;- p.adjust(raw_p_values, method = pcorr)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">109</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    out &lt;- cbind(raw_p_values, adjusted_p_values)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">110</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(out) &lt;- c("raw_p_values", "adjusted_p_values")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(out)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">112</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">113</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">114</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate p-values using label shuffling.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">115</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">116</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A \code{matrix} with the splicing diversity values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">117</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Character vector with an equal length to the number of columns</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">118</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' in the input dataset, specifying the category of each sample.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">119</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param control Name of the control sample category, defined in the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">120</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{samples} vector, e.g. \code{control = 'Normal'} or \code{control =</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">121</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' 'WT'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">122</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param method Method to use for calculating the average splicing diversity</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">123</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' value in a condition. Can be \code{'mean'} or \code{'median'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">124</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param randomizations The number of random shuffles.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">125</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pcorr P-value correction method applied to the results, as defined in</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">126</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the \code{p.adjust()} function.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">127</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired Logical; if \code{TRUE} perform a paired permutation scheme</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">128</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   (default: \code{FALSE}). When paired is \code{TRUE}, permutations</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">129</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   should preserve pairing between samples; the function currently permutes</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">130</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   sample labels and therefore paired analyses are only meaningful when the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">131</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   caller has arranged \code{samples} accordingly.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">132</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired_method Character; method for paired permutations. One of</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">133</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   \code{'swap'} (randomly swap labels within pairs) or \code{'signflip'}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">134</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   (perform sign-flip permutations; can enumerate all 2^n_pairs combinations</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">135</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   for an exact test when \code{randomizations = 0} or \code{randomizations &gt;= 2^n_pairs}).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">136</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return Raw and corrected p-values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">137</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @details The permutation p-values are computed two-sided as the proportion</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' of permuted log2 fold-changes at least as extreme as the observed value,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">139</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' with a pseudocount added: (count + 1) / (n_perm + 1).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">140</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import stats</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">141</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @note The permutation test returns two-sided empirical p-values using a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">142</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' pseudocount to avoid zero p-values for small numbers of permutations. See</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">143</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the function documentation for details.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">144</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">label_shuffling &lt;- function(x, samples, control, method, randomizations = 100, pcorr = "BH",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">145</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paired = FALSE, paired_method = c("swap", "signflip")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paired_method &lt;- match.arg(paired_method)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">147</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # observed log2 fold changes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">148</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    log2_fc &lt;- calculate_fc(x, samples, control, method)[, 4]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">149</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">150</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # build permutation/null distribution of log2 fold changes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">151</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (isTRUE(paired)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">152</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        perm_mat &lt;- .tsenat_permute_paired(x = x, samples = samples, control = control,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">153</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            method = method, randomizations = randomizations, paired_method = paired_method)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">154</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">155</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        permuted &lt;- replicate(randomizations, calculate_fc(x, sample(samples), control,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">156</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            method), simplify = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">157</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # each element is a data.frame/matrix; extract log2_fold_change column</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">158</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # (4th column)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">159</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        perm_mat &lt;- vapply(permuted, function(z) as.numeric(z[, 4]), numeric(nrow(x)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">161</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">162</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # compute two-sided permutation p-value with pseudocount: (count &gt;= |obs| +</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">163</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # 1) / (n_perm + 1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">164</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    raw_p_values &lt;- vapply(seq_len(nrow(perm_mat)), function(i) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">165</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        obs &lt;- log2_fc[i]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">166</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        nulls &lt;- perm_mat[i, ]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">167</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.na(obs) || all(is.na(nulls))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">168</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">169</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">170</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        nulls_non_na &lt;- nulls[!is.na(nulls)]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">171</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        n_non_na &lt;- length(nulls_non_na)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">172</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (n_non_na == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">173</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">174</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">175</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        cnt &lt;- sum(abs(nulls_non_na) &gt;= abs(obs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">176</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pval &lt;- (cnt + 1)/(n_non_na + 1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">177</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(pval)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">178</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }, numeric(1))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">179</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">180</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    adjusted_p_values &lt;- p.adjust(raw_p_values, method = pcorr)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">181</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    out &lt;- cbind(raw_p_values, adjusted_p_values)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">182</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(out) &lt;- c("raw_p_values", "adjusted_p_values")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">183</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(out)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">184</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">185</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">186</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">187</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Run a differential test by name: Wilcoxon or label-shuffle</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">188</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">189</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Thin wrapper that selects between `wilcoxon()` and `label_shuffling()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">190</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">191</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A matrix with splicing diversity values (rows = features).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">192</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples Character vector of sample group labels (length = ncol(x)).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">193</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param control Name of the control group (required for label-shuffle).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">194</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param method Character; one of `'wilcoxon'` or `'shuffle'`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">195</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param fc_method Character; aggregation method used by the permutation</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">196</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   test when `method = 'shuffle'` ('mean' or 'median').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">197</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired Logical passed to `wilcoxon()` when using the Wilcoxon test.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">198</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param exact Logical passed to `wilcoxon()` to request exact p-values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">199</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param randomizations Integer number of permutations for `label_shuffling()`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">200</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pcorr P-value adjustment method (passed to `p.adjust`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">201</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param seed Integer seed used to make permutations reproducible (default</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">202</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   123). The function sets a temporary RNG seed via `withr::local_seed(seed)`</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">203</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   before running `label_shuffling()` when `method = 'shuffle'`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">204</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A two-column matrix with raw and adjusted p-values (as returned by</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">205</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   the underlying functions).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">206</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">207</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">208</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' mat &lt;- matrix(rnorm(20), nrow = 5)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">209</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' samples &lt;- rep(c('A', 'B'), length.out = ncol(mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">210</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' test_differential(mat, samples, control = 'A', method = 'wilcoxon')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">211</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">212</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired_method Character; forwarded to `label_shuffling()` when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">213</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `method = 'shuffle'`. See `label_shuffling()` for details.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">214</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">test_differential &lt;- function(x, samples, control = NULL, method = c("wilcoxon",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">215</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    "shuffle"), fc_method = "mean", paired = FALSE, exact = FALSE, randomizations = 100,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">216</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pcorr = "BH", seed = 123L, paired_method = c("swap", "signflip")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">217</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paired_method &lt;- match.arg(paired_method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">218</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    method &lt;- match.arg(method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">219</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "wilcoxon") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">220</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(wilcoxon(x, samples, pcorr = pcorr, paired = paired, exact = exact))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">221</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">222</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">223</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # shuffle / permutation-based test</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">224</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(control) || !nzchar(control)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">225</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("`control` must be provided when method = 'shuffle'", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">226</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">227</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(seed)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">228</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.numeric(seed) || length(seed) != 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">229</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("`seed` must be a single numeric value", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">230</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">231</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # Use withr::local_seed to temporarily set RNG state for reproducible</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">232</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # permutation tests without permanently modifying the global RNG.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">233</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        withr::local_seed(as.integer(seed))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">234</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">235</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(label_shuffling(x, samples, control, fc_method, randomizations = randomizations,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">236</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pcorr = pcorr, paired = paired, paired_method = paired_method))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">237</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/filter_se.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">## Filter SummarizedExperiment by low-expression transcripts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Filter transcripts in a `SummarizedExperiment` by minimum count/sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Subset a `SummarizedExperiment` to keep transcripts with more than</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `min_count` reads in strictly more than `min_samples` samples. The</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' function updates assays, `rowData`, and relevant entries in</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `metadata()` (for example `readcounts` and `tx2gene`) so downstream</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' helpers receive a consistent object.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` object containing transcript-level</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   assays (rows = transcripts, columns = samples).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param min_count Integer threshold for per-sample counts (default 5L).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param min_samples Integer minimum number of samples exceeding</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   `min_count` required to keep a transcript (default 5L).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name or index of the assay to use for filtering</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   (default: 'counts'). If not present, the first assay is used.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose Logical; print before/after counts when TRUE.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A filtered `SummarizedExperiment`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' mat &lt;- matrix(c(0, 6, 7, 2, 8, 9), nrow = 3, dimnames = list(paste0('tx', 1:3), paste0('S', 1:2)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- SummarizedExperiment::SummarizedExperiment(assays = list(counts = mat))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' filt &lt;- filter_se(se, min_count = 5, min_samples = 1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' class(filt)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">filter_se &lt;- function(se, min_count = 5L, min_samples = 5L, assay_name = "counts",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    verbose = TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is(se, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("'se' must be a SummarizedExperiment", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assays_list &lt;- SummarizedExperiment::assays(se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.character(assay_name) &amp;&amp; assay_name %in% names(assays_list)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assay_mat &lt;- as.matrix(assays_list[[assay_name]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (is.numeric(assay_name) &amp;&amp; assay_name &gt;= 1 &amp;&amp; assay_name &lt;= length(assays_list)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">35</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assay_mat &lt;- as.matrix(assays_list[[assay_name]])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # fallback to first assay</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assay_mat &lt;- as.matrix(assays_list[[1]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        warning("Requested assay not found; using first assay.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">40</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">42</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(assay_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">43</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Assay data must be numeric.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">44</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">45</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">46</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    before &lt;- nrow(assay_mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Cap min_samples to the number of available samples to avoid impossible</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # thresholds on small datasets and use an inclusive threshold so callers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # requesting e.g. 'min_samples = ncol' keep rows present in all samples.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    eff_min_samples &lt;- min(as.integer(min_samples), ncol(assay_mat))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tokeep &lt;- rowSums(assay_mat &gt; min_count) &gt;= eff_min_samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    after &lt;- sum(tokeep)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">55</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        message(sprintf("Transcripts: before = %d, after = %d", before, after))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (after == 0L) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        warning("Filtering removed all transcripts; returning empty SummarizedExperiment.",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">62</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # subset all assays</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    new_assays &lt;- S4Vectors::SimpleList(lapply(assays_list, function(a) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.matrix(a) || is.data.frame(a)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            as.matrix(a)[tokeep, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">67</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">68</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">69</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">70</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    names(new_assays) &lt;- names(assays_list)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">72</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">73</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # subset rowData if present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rd &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(SummarizedExperiment::rowData(se)) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rd &lt;- SummarizedExperiment::rowData(se)[tokeep, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">78</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">79</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # subset metadata readcounts and tx2gene if present</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    md &lt;- S4Vectors::metadata(se)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(md$readcounts) &amp;&amp; is.matrix(md$readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">82</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md$readcounts &lt;- as.matrix(md$readcounts)[tokeep, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">83</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.null(md$tx2gene) &amp;&amp; is.data.frame(md$tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txmap &lt;- md$tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        txcol &lt;- colnames(txmap)[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md$tx2gene &lt;- txmap[txmap[[txcol]] %in% rownames(assay_mat)[tokeep], , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">88</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">89</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">90</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # construct new SE with same colData</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    new_se &lt;- SummarizedExperiment::SummarizedExperiment(assays = new_assays, rowData = if (!is.null(rd)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rd</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">93</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">94</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        S4Vectors::DataFrame()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }, colData = SummarizedExperiment::colData(se), metadata = c(S4Vectors::metadata(se),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        list(filtered = list(min_count = min_count, min_samples = min_samples))))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">97</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # attach updated metadata pieces</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    S4Vectors::metadata(new_se)$readcounts &lt;- if (!is.null(md$readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md$readcounts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">103</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    S4Vectors::metadata(new_se)$tx2gene &lt;- if (!is.null(md$tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md$tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">108</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">109</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">110</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(new_se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">111</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/difference_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Helper utilities for calculate_difference</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_calculate_difference_partition &lt;- function(df, samples, control, method,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    test, pcorr, randomizations, verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (ncol(df) - 1 != length(samples))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">6</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Column count doesn't match length(samples).", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    uniq_groups &lt;- unique(as.character(samples))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(uniq_groups) &gt; 2)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">9</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("More than two conditions; provide exactly two.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">23<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(uniq_groups) &lt; 2)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">11</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Fewer than two conditions; provide exactly two.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">12</td>
                    <td class="coverage">19<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(control %in% uniq_groups))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">13</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Control sample type not found in samples.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">15</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    case_label &lt;- setdiff(uniq_groups, control)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">16</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    groups &lt;- c(case_label, control)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">18</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(method %in% c("mean", "median")))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">19</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Invalid method; see ?calculate_difference.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">20</td>
                    <td class="coverage">14<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(test %in% c("wilcoxon", "shuffle")))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">21</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Invalid test method; see ?calculate_difference.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">22</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    valid_pcorr &lt;- c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">23</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "none")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">24</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(pcorr %in% valid_pcorr))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">25</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Invalid p-value correction; see ?calculate_difference.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tab &lt;- table(samples)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (test == "wilcoxon") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (randomizations != 100 &amp;&amp; verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            message("'randomizations' ignored for wilcoxon.")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (any(tab &lt; 3) || sum(tab) &lt; 8)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            warning("Low sample size for wilcoxon.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (test == "shuffle") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">35</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (sum(tab) &lt;= 5)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            warning("Low sample size for label shuffling.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (sum(tab) &gt; 5 &amp;&amp; sum(tab) &lt; 10)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">38</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            warning("Label shuffling may be unreliable.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">39</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">40</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx_case &lt;- which(samples == groups[1])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">42</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx_control &lt;- which(samples == groups[2])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx1 &lt;- idx_case</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx2 &lt;- idx_control</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">45</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">46</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$cond_1 &lt;- rowSums(!is.na(df[, idx1 + 1, drop = FALSE]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">47</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$cond_2 &lt;- rowSums(!is.na(df[, idx2 + 1, drop = FALSE]))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (test == "wilcoxon") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        keep_mask &lt;- (df$cond_1 &gt;= 3 &amp; df$cond_2 &gt;= 3 &amp; (df$cond_1 + df$cond_2) &gt;=</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            8)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        keep_mask &lt;- (df$cond_1 + df$cond_2) &gt;= 5</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_keep &lt;- df[keep_mask, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_small &lt;- df[!keep_mask, , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">58</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(df = df, samples = samples, groups = groups, idx1 = idx1, idx2 = idx2, df_keep = df_keep,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        df_small = df_small)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">62</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Helpers for calculate_fc</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_aggregate_fc_values &lt;- function(x, samples, method, control) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (method == "mean") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">543<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        value &lt;- aggregate(t(x), by = list(samples), mean, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (method == "median") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        value &lt;- aggregate(t(x), by = list(samples), median, na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">69</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">70</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Invalid method; must be 'mean' or 'median'")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">72</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sorted &lt;- value[value$Group.1 != control, ]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sorted[2, ] &lt;- value[value$Group.1 == control, ]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value &lt;- t(sorted[, -1])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value[is.na(value[, 1]), c(1)] &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value[is.na(value[, 2]), c(2)] &lt;- NA</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(list(value = value, sorted = sorted))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">79</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">80</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">81</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_apply_pseudocount &lt;- function(value, pseudocount) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">82</td>
                    <td class="coverage">549<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(pseudocount) || length(pseudocount) != 1) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">83</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pseudocount &lt;- 0</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">84</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">549<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (pseudocount &lt;= 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pos_vals &lt;- value[!is.na(value) &amp; value &gt; 0]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">547<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(pos_vals) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">545<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pc &lt;- min(pos_vals, na.rm = TRUE)/2</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">89</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            pc &lt;- 1e-06</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">91</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">92</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        pc &lt;- pseudocount</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">549<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    replace_idx &lt;- !is.na(value) &amp; value &lt;= 0</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">549<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    value[replace_idx] &lt;- pc</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">549<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(value)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">99</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Paired permutation helpers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_permute_paired &lt;- function(x, samples, control, method, randomizations, paired_method) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    ncols &lt;- ncol(x)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (ncols%%2 != 0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Paired permutation requires an even number of samples and paired column ordering",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">106</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    npairs &lt;- ncols/2</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (paired_method == "swap") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">108</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        perm_mat &lt;- matrix(NA_real_, nrow = nrow(x), ncol = randomizations)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">109</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        for (r in seq_len(randomizations)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">110</td>
                    <td class="coverage">110<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            swap &lt;- sample(c(TRUE, FALSE), size = npairs, replace = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">110<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            perm_samples &lt;- samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">112</td>
                    <td class="coverage">110<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            for (p in seq_len(npairs)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">113</td>
                    <td class="coverage">220<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (swap[p]) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">114</td>
                    <td class="coverage">123<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  i1 &lt;- (p - 1) * 2 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">115</td>
                    <td class="coverage">123<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  i2 &lt;- i1 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">116</td>
                    <td class="coverage">123<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  perm_samples[c(i1, i2)] &lt;- perm_samples[c(i2, i1)]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">117</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">118</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">119</td>
                    <td class="coverage">110<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            df_perm &lt;- calculate_fc(x, perm_samples, control, method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">120</td>
                    <td class="coverage">110<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            perm_mat[, r] &lt;- as.numeric(df_perm[, 4])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">121</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">122</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(perm_mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">123</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (paired_method == "signflip") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">124</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        total_comb &lt;- 2^npairs</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (randomizations &lt;= 0 || randomizations &gt;= total_comb) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">126</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            combos &lt;- expand.grid(rep(list(c(0, 1)), npairs))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">127</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            nrep &lt;- nrow(combos)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">128</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            perm_mat &lt;- matrix(NA_real_, nrow = nrow(x), ncol = nrep)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">129</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            for (r in seq_len(nrep)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">130</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                swap &lt;- as.logical(as.integer(combos[r, ]))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">131</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                perm_samples &lt;- samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">132</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                for (p in seq_len(npairs)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">133</td>
                    <td class="coverage">40<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  if (swap[p]) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">134</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    i1 &lt;- (p - 1) * 2 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">135</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    i2 &lt;- i1 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">136</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    perm_samples[c(i1, i2)] &lt;- perm_samples[c(i2, i1)]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">137</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">139</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                df_perm &lt;- calculate_fc(x, perm_samples, control, method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">140</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                perm_mat[, r] &lt;- as.numeric(df_perm[, 4])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">141</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">142</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(perm_mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">143</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">144</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            perm_mat &lt;- matrix(NA_real_, nrow = nrow(x), ncol = randomizations)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">145</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            for (r in seq_len(randomizations)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                swap &lt;- sample(c(TRUE, FALSE), size = npairs, replace = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">147</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                perm_samples &lt;- samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">148</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                for (p in seq_len(npairs)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">149</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  if (swap[p]) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    i1 &lt;- (p - 1) * 2 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">151</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    i2 &lt;- i1 + 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">152</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    perm_samples[c(i1, i2)] &lt;- perm_samples[c(i2, i1)]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">153</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">154</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">155</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                df_perm &lt;- calculate_fc(x, perm_samples, control, method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">156</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                perm_mat[, r] &lt;- as.numeric(df_perm[, 4])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">157</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">158</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(perm_mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">159</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">161</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/diversity_functions.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># # Minimalized diversity functions: only Tsallis entropy retained</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate Tsallis entropy for a vector of transcript-level</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' expression values of one gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x Vector of (non-negative) expression values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param q Tsallis entropy parameter (q &gt; 0). Scalar or numeric vector</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (default: 2).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param norm Logical; if TRUE, normalize entropy by its theoretical maximum</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (values in [0,1]).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param what Which quantity to return: 'S' (Tsallis entropy), 'D' (Hill</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' numbers), or 'both'.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param log_base Base of the logarithm used for Shannon limits and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' normalization</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (default: \code{exp(1)}).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return For `what = 'S'` or `what = 'D'`: a numeric vector</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (named when length(q) &gt; 1). For `what = 'both'`: a list with</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' components `$S` and `$D`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @details</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Implements S_q = (1 - sum p^q)/(q - 1) and D_q = (sum p^q)^(1/(1-q)).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Uses the q-&gt;1 limits (Shannon entropy and its exponential). Natural</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' logarithms are used for the q-&gt;1 limit and for normalization.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' x &lt;- c(10, 5, 0)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' calculate_tsallis_entropy(x, q = c(0.5, 1, 2), norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_tsallis_entropy &lt;- function(x, q = 2, norm = TRUE, what = c("S", "D", "both"),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    log_base = exp(1)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">401<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    what &lt;- match.arg(what)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">401<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(q)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("q must be numeric.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">400<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (any(q &lt;= 0)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("q must be greater than 0.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">395<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("x must be numeric")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">39</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">394<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    n &lt;- length(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # If all counts sum to zero, return NA; allow single-element vectors to</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # proceed</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">394<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (sum(x, na.rm = TRUE) &lt;= 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (what == "both") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(list(S = rep(NA_real_, length(q)), D = rep(NA_real_, length(q))))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">47</td>
                    <td class="coverage">30<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(rep(NA_real_, length(q)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    p &lt;- x/sum(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tol &lt;- sqrt(.Machine$double.eps)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    S_vec &lt;- .tsenat_calc_S(p = p, q = q, tol = tol, n = n, log_base = log_base,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        norm = norm)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    D_vec &lt;- .tsenat_calc_D(p = p, q = q, tol = tol, log_base = log_base)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">363<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (what == "S") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">348<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        out &lt;- S_vec</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">348<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(q) &gt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">231<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            names(out) &lt;- paste0("q=", q)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">348<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(q) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">117<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(unname(out))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">231<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(out)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">66</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (what == "D") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        out &lt;- D_vec</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(q) &gt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            names(out) &lt;- paste0("q=", q)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(q) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(unname(out))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">74</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(out)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # both</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">78</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    names(S_vec) &lt;- paste0("q=", q)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">79</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    names(D_vec) &lt;- paste0("q=", q)</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">80</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(list(S = S_vec, D = D_vec))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">81</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/diversity_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Helpers for Tsallis entropy calculations</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_calc_S &lt;- function(p, q, tol, n, log_base, norm) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">4</td>
                    <td class="coverage">365<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    vapply(q, function(qi) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">839<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (abs(qi - 1) &lt; tol) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">6</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            sh &lt;- -sum(ifelse(p &gt; 0, p * log(p, base = log_base), 0))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (norm) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">131<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                sh &lt;- sh/log(n, base = log_base)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(sh)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">12</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ts &lt;- (1 - sum(p^qi))/(qi - 1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">13</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (norm) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">14</td>
                    <td class="coverage">698<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                max_ts &lt;- (1 - n^(1 - qi))/(qi - 1)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">15</td>
                    <td class="coverage">698<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                ts &lt;- ts/max_ts</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">17</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(ts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">19</td>
                    <td class="coverage">365<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }, numeric(1))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_calc_D &lt;- function(p, q, tol, log_base) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">23</td>
                    <td class="coverage">365<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    vapply(q, function(qi) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">24</td>
                    <td class="coverage">839<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (abs(qi - 1) &lt; tol) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">25</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            sh &lt;- -sum(ifelse(p &gt; 0, p * log(p, base = log_base), 0))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">26</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            D1 &lt;- (log_base)^(sh)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">134<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(D1)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            spq &lt;- sum(p^qi)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            Dq &lt;- spq^(1/(1 - qi))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">705<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            return(Dq)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">365<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }, numeric(1))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Input preparation</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_diversity_input &lt;- function(x, genes = NULL, tpm = FALSE, assayno = 1,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">46<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(is.matrix(x) || is.data.frame(x) || is.list(x) || is(x, "DGEList") || is(x,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">46<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "RangedSummarizedExperiment") || is(x, "SummarizedExperiment"))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Input data type is not supported! Please use ?calculate_diversity to see the possible arguments and details.",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">43</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">44<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is(x, "data.frame")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x &lt;- as.matrix(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">48</td>
                    <td class="coverage">44<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (tpm == TRUE &amp;&amp; !is.list(x) &amp;&amp; verbose == TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        message("Note: tpm as a logical argument is only interpreted in case of",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            " tximport lists.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">44<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    se_assay_mat &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">44<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.list(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(x) == 4 &amp;&amp; "counts" %in% names(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">56</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (tpm == FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                x &lt;- as.matrix(x$counts)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">58</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">59</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (tpm == TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">60</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                x &lt;- as.matrix(x$abundance)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else if (is(x, "DGEList")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x &lt;- as.matrix(x$counts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (verbose == TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                message("Note: calculate_diversity methods are only applicable",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  " if your DGEList contains transcript-level expression", " data.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">67</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (tpm == TRUE &amp;&amp; verbose == TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                message("Note: tpm as a logical argument is only interpreted", " in case of tximport lists.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">70</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("The package cannot find any expression data in your input.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">73</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">74</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">75</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">76</td>
                    <td class="coverage">42<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is(x, "RangedSummarizedExperiment") || is(x, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        md &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        try(md &lt;- S4Vectors::metadata(x), silent = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!is.null(md) &amp;&amp; !is.null(md$readcounts)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            se_assay_mat &lt;- as.matrix(md$readcounts)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x &lt;- se_assay_mat</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            assays_len &lt;- length(SummarizedExperiment::assays(x))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.numeric(assayno) || assays_len &lt; assayno) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("Please provide a valid assay number.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">86</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            se_assay_mat &lt;- as.matrix(SummarizedExperiment::assays(x)[[assayno]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x &lt;- se_assay_mat</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">89</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (is.null(genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (exists("se_assay_mat") &amp;&amp; !is.null(md) &amp;&amp; !is.null(md$tx2gene) &amp;&amp;</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                is.data.frame(md$tx2gene)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                txmap &lt;- md$tx2gene</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">94</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                tx_col &lt;- if ("Transcript" %in% colnames(txmap)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  "Transcript"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">96</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  colnames(txmap)[1]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                gene_col &lt;- if ("Gen" %in% colnames(txmap)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  "Gen"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  colnames(txmap)[2]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">103</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                genes &lt;- as.character(txmap[[gene_col]][match(rownames(se_assay_mat),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  txmap[[tx_col]])])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">106</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                rownames(x) &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">107</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">108</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                genes &lt;- rownames(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">109</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # keep transcript rownames available for downstream metadata</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">110</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                # (we will store original transcript-level counts separately)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">111</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                rownames(x) &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">112</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">113</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">114</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (is.null(genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">115</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("Please construct a valid gene set for your ", "SummarizedExperiment.",</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">116</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">117</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">118</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">119</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">120</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">121</td>
                    <td class="coverage">39<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(x = x, genes = genes, se_assay_mat = se_assay_mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">122</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/calculate_difference.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate splicing diversity changes between two conditions.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A \code{SummarizedExperiment} with splicing diversity values for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' each gene in each sample or a \code{data.frame} with gene names in the  first</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' column and splicing diversity values for each sample in additional  columns.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param samples A vector of length one, specifying the column name of the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{colData} annotation column from the \code{SummarizedExperiment}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' object, that should be used as the category column or a character vector</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' with an equal length to the number of columns in the input dataset,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' specifying the category of each sample in the case of a \code{data.frame}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' input.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param control Name of the control sample category, defined in the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{samples} vector, e.g. \code{control = 'Normal'} or \code{control =</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' 'WT'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param method Method to use for calculating the average splicing diversity</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' value in a condition. Can be \code{'mean'} or \code{'median'}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param test Method to use for p-value calculation: use \code{'wilcoxon'} for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Wilcoxon rank sum test or \code{'shuffle'} for a label shuffling test.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param randomizations Number of random shuffles, used for the label shuffling</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' test (default = 100).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pcorr P-value correction method applied to the Wilcoxon rank sum test</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' or label shuffling test results, as defined in the \code{p.adjust}  function.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assayno An integer value. In case of multiple assays in a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{SummarizedExperiment} input, the argument specifies the assay number</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' to use for difference calculations.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose If \code{TRUE}, the function will print additional diagnostic</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' messages.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pseudocount Numeric scalar. Passed to \code{calculate_fc} and used to</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' add a small value to non-positive group summaries before computing</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' differences and log2 fold-changes. Default \code{1e-6}. Rows excluded for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' low sample counts remain \code{NA}.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">31</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired Logical; if `TRUE`, run paired versions of tests when</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   supported (default: `FALSE`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param exact Logical; passed to the Wilcoxon test to request exact p-values</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'   when supported (default: `FALSE`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A \code{data.frame} with the mean or median values of splicing</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">36</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' diversity across sample categories and all samples, log2(fold change) of  the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' two different conditions, raw and corrected p-values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import methods</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">39</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom SummarizedExperiment SummarizedExperiment assays assay colData</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">40</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @details The function calculates diversity changes between two sample</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' conditions. It uses the output of the diversity calculation function, which</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">43</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' is a \code{SummarizedExperiment} object of splicing diversity values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">44</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Additionally, it can use a \code{data.frame} as input, where the first column</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">45</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' contains gene names, and all additional columns contain splicing diversity</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' values for each sample. A vector of sample conditions also serves as input,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' used for aggregating the samples by condition.   It calculates the mean or</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' median of the splicing diversity data per sample  condition, the difference</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">49</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' of these values and the log2 fold change of the two  conditions. Furthermore,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the user can select a statistical method to  calculate the significance of</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the changes. The p-values and adjusted p-values  are calculated using a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Wilcoxon sum rank test or label shuffling test.   The function will exclude</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' genes of low sample size from the significance  calculation, depending on</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' which statistical test is applied.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' x &lt;- data.frame(Genes = letters[seq_len(10)], matrix(runif(80), ncol = 8))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' samples &lt;- c(rep('Healthy', 4), rep('Pathogenic', 4))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">58</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' calculate_difference(x, samples,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">59</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     control = 'Healthy', method = 'mean', test =</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">60</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'         'wilcoxon'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">62</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_difference &lt;- function(x, samples = NULL, control, method = "mean", test = "wilcoxon",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    randomizations = 100, pcorr = "BH", assayno = 1, verbose = TRUE, paired = FALSE,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    exact = FALSE, pseudocount = 0) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">65</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # internal small helpers (kept here to avoid adding new files)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .tsenat_prepare_df &lt;- function(x, samples, assayno) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">34<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (inherits(x, "RangedSummarizedExperiment") || inherits(x, "SummarizedExperiment")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">68</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # allow samples to be NULL (use default 'sample_type' col)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (is.null(samples)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if ("sample_type" %in% colnames(SummarizedExperiment::colData(x))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  samples_col &lt;- "sample_type"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">72</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  stop("When providing a SummarizedExperiment, supply 'samples' as a colData column name or call map_metadata() to populate 'sample_type'",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                    call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">75</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                if (length(samples) != 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                  stop("'samples' must be a single colData column.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">79</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                }</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">80</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                samples_col &lt;- samples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">81</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">82</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            samples_vec &lt;- SummarizedExperiment::colData(x)[[samples_col]]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (!is.numeric(assayno) || length(SummarizedExperiment::assays(x)) &lt;</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                assayno) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("Invalid 'assayno'.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">86</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">87</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            df &lt;- as.data.frame(SummarizedExperiment::assays(x)[[assayno]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            genes &lt;- rownames(df)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            df &lt;- cbind(genes = genes, df)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            list(df = df, samples = samples_vec)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">91</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">30<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            df &lt;- as.data.frame(x)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">30<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            list(df = df, samples = samples)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">95</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">96</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .tsenat_sample_matrix &lt;- function(dfr) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">98</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        as.matrix(dfr[, -c(1, ncol(dfr) - 1, ncol(dfr)), drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">99</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Validate input container Reject matrices explicitly (tests expect this</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">102</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # error for matrix input)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.matrix(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">4<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Input type unsupported; see ?calculate_difference.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">105</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">106</td>
                    <td class="coverage">34<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(is.data.frame(x) || inherits(x, "RangedSummarizedExperiment") || inherits(x,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">34<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "SummarizedExperiment"))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">108</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Input data type not supported; see ?calculate_difference.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">109</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">110</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">111</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # prepare data.frame and sample vector (handles SummarizedExperiment)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">112</td>
                    <td class="coverage">34<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pd &lt;- .tsenat_prepare_df(x, samples, assayno)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">113</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- pd$df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">114</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    samples &lt;- pd$samples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">115</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">116</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Partition and validate inputs</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">117</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    part &lt;- .tsenat_calculate_difference_partition(df = df, samples = samples, control = control,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">118</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        method = method, test = test, pcorr = pcorr, randomizations = randomizations,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">119</td>
                    <td class="coverage">31<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        verbose = verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">120</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- part$df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">121</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    samples &lt;- part$samples</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">122</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    groups &lt;- part$groups</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">123</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx1 &lt;- part$idx1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">124</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx2 &lt;- part$idx2</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">125</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_keep &lt;- part$df_keep</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">126</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df_small &lt;- part$df_small</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">127</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">128</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_list &lt;- list()</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">129</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">130</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Helper to extract the numeric matrix of sample columns (keeps original</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">131</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # order)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">132</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_matrix &lt;- .tsenat_sample_matrix</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">133</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">134</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(df_keep) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">135</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (nrow(df_small) &gt; 0 &amp;&amp; verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">136</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            message(sprintf("Note: %d genes excluded due to low sample counts.",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">137</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                nrow(df_small)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">138</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">139</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        ymat &lt;- sample_matrix(df_keep)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">140</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # p-value calculation</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">141</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (test == "wilcoxon") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">142</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ptab &lt;- wilcoxon(ymat, samples, pcorr = pcorr, paired = paired, exact = exact)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">143</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # wilcoxon should return a data.frame of p-values named</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">144</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            # appropriately</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">145</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">146</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            ptab &lt;- label_shuffling(ymat, samples, control, method, randomizations = randomizations,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">147</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                pcorr = pcorr, paired = paired)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">148</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">149</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_list$tested &lt;- data.frame(genes = df_keep[, 1], calculate_fc(ymat,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">150</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            samples, control, method, pseudocount = pseudocount), ptab, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">151</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">152</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">153</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(df_small) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">154</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        small_mat &lt;- sample_matrix(df_small)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">155</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_list$small &lt;- data.frame(genes = df_small[, 1], calculate_fc(small_mat,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">156</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            samples, control, method, pseudocount = pseudocount), raw_p_values = NA,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">157</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            adjusted_p_values = NA, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">158</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">159</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">160</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Combine results preserving tested rows first</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">161</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(result_list) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">162</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(data.frame())</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">163</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">164</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    res &lt;- do.call(rbind, result_list)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">165</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rownames(res) &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">166</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!("log2_fold_change" %in% colnames(res)) &amp;&amp; ("log2FC" %in% colnames(res))) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">167</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        res$log2_fold_change &lt;- res$log2FC</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">168</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">169</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(res)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">170</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">171</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">172</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">173</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Linear-model interaction test for Tsallis entropy   For each gene, fit a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">174</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' linear model of the form `entropy ~ q * group` and  extract the p-value for</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">175</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' the interaction term (whether the effect of `q`  differs between groups). The</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">176</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' function expects a `SummarizedExperiment`  produced by</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">177</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `calculate_diversity()` when multiple `q` values have been  computed (column</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">178</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' names contain `_q=`).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">179</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param se A `SummarizedExperiment` containing a `diversity` assay produced</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">180</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' by `calculate_diversity(..., q = &lt;vector&gt;)`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">181</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param sample_type_col Optional column name in `colData(se)` that contains</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">182</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' a grouping factor for samples (character). If `NULL`, the function will</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">183</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' attempt to infer group from column names (suffix `_N` interpreted as</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">184</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' 'Normal').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">185</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param min_obs Minimum number of non-NA observations required to fit a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">186</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' model for a gene (default: 10).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">187</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param method Modeling method to use for interaction testing: one of</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">188</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{c('linear', 'lmm', 'gam', 'fpca')} (default: 'linear').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">189</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param pvalue Type of p-value to compute for linear mixed models: one of</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">190</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' \code{c('satterthwaite', 'lrt', 'both')} (default: 'satterthwaite').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">191</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param subject_col Optional column name in `colData(se)` that contains</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">192</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' subject/individual identifiers for paired or repeated-measures designs</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">193</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (character). If provided with `method = 'lmm'`, used as random effect.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">194</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param paired Logical; whether samples are paired (default: FALSE).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">195</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param nthreads Number of threads (mc.cores) to use for parallel processing</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">196</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (default: 1).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">197</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assay_name Name of the assay in the SummarizedExperiment to use</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">198</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (default: 'diversity').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">199</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose Logical; whether to print progress messages during execution</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">200</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' (default: FALSE).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">201</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A data.frame with columns `gene`, `p_interaction`, and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">202</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `adj_p_interaction`, ordered by ascending `p_interaction`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">203</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">204</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">205</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data('tcga_brca_luma_dataset', package = 'TSENAT')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">206</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">207</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">208</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = c(0.1, 1), norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">209</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' # Provide a minimal sample-type mapping so the example runs during checks</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">210</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' SummarizedExperiment::colData(se) &lt;- S4Vectors::DataFrame(</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">211</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     sample_type = rep(c('Normal', 'Tumor'), length.out = ncol(se)),</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">212</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'     row.names = colnames(se)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">213</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' )</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">214</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' calculate_lm_interaction(se, sample_type_col = 'sample_type')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">215</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_lm_interaction &lt;- function(se, sample_type_col = NULL, min_obs = 10, method = c("linear",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">216</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    "lmm", "gam", "fpca"), pvalue = c("satterthwaite", "lrt", "both"), subject_col = NULL,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">217</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paired = FALSE, nthreads = 1, assay_name = "diversity", verbose = FALSE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">218</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    method &lt;- match.arg(method)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">219</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    pvalue &lt;- match.arg(pvalue)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">220</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (verbose) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">221</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        message("[calculate_lm_interaction] method=", method)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">222</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">223</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # internal flags: keep these internal to avoid documenting them in Rd</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">224</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    suppress_lme4_warnings &lt;- TRUE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">225</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    progress &lt;- FALSE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">226</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">227</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("SummarizedExperiment required")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">228</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">229</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">230</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    mat &lt;- SummarizedExperiment::assay(se, assay_name)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">231</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">232</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(sprintf("Assay '%s' not found in SummarizedExperiment", assay_name))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">233</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">234</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">235</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_q &lt;- colnames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">236</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(sample_q) || length(sample_q) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">237</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("No column names found on diversity assay")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">238</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">239</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">240</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # parse sample names and q values from column names like 'Sample_q=0.01'</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">241</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_names &lt;- sub("_q=.*", "", sample_q)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">242</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    has_q &lt;- grepl("_q=", sample_q)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">243</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!any(has_q)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">244</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Could not parse q values; expected '_q=' in column names", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">245</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">246</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!all(has_q)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">247</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Some column names are missing '_q='; ensure all diversity columns include a q value",</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">248</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">249</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">250</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    q_vals &lt;- as.numeric(sub(".*_q=", "", sample_q))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">251</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">252</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # determine group for each sample</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">253</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_type_in_coldata &lt;- !is.null(sample_type_col) &amp;&amp; sample_type_col %in% colnames(SummarizedExperiment::colData(se))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">254</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (sample_type_in_coldata) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">255</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        st &lt;- as.character(SummarizedExperiment::colData(se)[, sample_type_col])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">256</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        names(st) &lt;- SummarizedExperiment::colData(se)$samples %||% colnames(mat)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">257</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # when user provides a sample_type_col, index by the sample names</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">258</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        # (strip q suffix)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">259</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        group_vec &lt;- unname(st[sample_names])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">260</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">261</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("No sample grouping found: please supply `sample_type_col` or map sample",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">262</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            "types into `colData(se)` before calling calculate_lm_interaction().",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">263</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">264</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">265</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">266</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (verbose &amp;&amp; progress) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">267</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        message("[calculate_lm_interaction] parsed samples and groups")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">268</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">269</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    all_results &lt;- list()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">270</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    fit_one &lt;- function(g) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">271</td>
                    <td class="coverage">65<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .tsenat_fit_one_interaction(g = g, se = se, mat = mat, q_vals = q_vals, sample_names = sample_names,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">272</td>
                    <td class="coverage">65<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            group_vec = group_vec, method = method, pvalue = pvalue, subject_col = subject_col,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">273</td>
                    <td class="coverage">65<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            paired = paired, min_obs = min_obs, verbose = verbose, suppress_lme4_warnings = suppress_lme4_warnings,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">274</td>
                    <td class="coverage">65<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            progress = progress)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">275</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">276</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">277</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nthreads &gt; 1 &amp;&amp; .Platform$OS.type == "unix") {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">278</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (!requireNamespace("parallel", quietly = TRUE)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">279</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stop("parallel package required for multi-threading")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">280</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">281</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        res_list &lt;- parallel::mclapply(rownames(mat), fit_one, mc.cores = nthreads)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">282</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">283</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        res_list &lt;- lapply(rownames(mat), fit_one)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">284</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">285</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    all_results &lt;- Filter(Negate(is.null), res_list)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">286</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">287</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(all_results) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">288</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(data.frame())</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">289</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">290</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    res &lt;- do.call(rbind, all_results)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">291</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    res$adj_p_interaction &lt;- stats::p.adjust(res$p_interaction, method = "BH")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">292</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Sort first by adjusted p-values, then by raw p-values for stable ordering</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">293</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    res &lt;- res[order(res$adj_p_interaction, res$p_interaction), , drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">294</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rownames(res) &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">295</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">296</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    .tsenat_report_fit_summary(res, verbose = verbose)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">297</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">298</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Return the result data.frame (do not attach to or return a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">299</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # SummarizedExperiment)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">300</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(res)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">301</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">302</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">303</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># small helper (replacement for `%||%`) to provide default when NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">304</td>
                    <td class="coverage">84<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">`%||%` &lt;- function(a, b) if (is.null(a)) b else a</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/calculate_diversity.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate Tsallis diversity per gene across samples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#'</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x A numeric matrix or data.frame of transcript-level expression</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' values (rows = transcripts, columns = samples), or a SummarizedExperiment-</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' like object.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param tpm Logical. If TRUE and `x` is a tximport-style list, use the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `$abundance` matrix instead of `$counts`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param genes Character vector assigning each transcript (row) to a gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Must have length equal to nrow(x) or the number of transcripts in `x`.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param norm Logical; if TRUE, normalize Tsallis entropy to [0,1] per gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param assayno Integer assay index to use when `x` is a SummarizedExperiment.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose Logical; print diagnostic messages when TRUE.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param q Numeric scalar or vector of Tsallis q values to evaluate (q &gt; 0).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' If length(q) &gt; 1, the result will contain separate columns per sample and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' q.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param what Which quantity to return: 'S' for Tsallis entropy or 'D' for Hill</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' numbers.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A \link[SummarizedExperiment]{SummarizedExperiment} with assay</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' `diversity` containing per-gene diversity values.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import methods</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @importFrom SummarizedExperiment SummarizedExperiment assays assay rowData</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' colData</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">23</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @export</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @examples</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' data('tcga_brca_luma_dataset', package = 'TSENAT')</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' rc &lt;- as.matrix(tcga_brca_luma_dataset[1:20, -1, drop = FALSE])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">27</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' gs &lt;- tcga_brca_luma_dataset$genes[1:20]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">28</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' se &lt;- calculate_diversity(rc, gs, q = 0.1, norm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">29</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' SummarizedExperiment::assay(se)[1:3, 1:3]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_diversity &lt;- function(x, genes = NULL, norm = TRUE, tpm = FALSE, assayno = 1,</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">31</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    verbose = FALSE, q = 2, what = c("S", "D")) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">32</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Normalize and validate input data, extract matrix and gene mapping</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">29<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    inp &lt;- .tsenat_prepare_diversity_input(x = x, genes = genes, tpm = tpm, assayno = assayno,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">34</td>
                    <td class="coverage">29<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        verbose = verbose)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">35</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x &lt;- inp$x</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    genes &lt;- inp$genes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    se_assay_mat &lt;- inp$se_assay_mat</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">38</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(x)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Input data  must be numeric!", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">41</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">26<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (any(is.na(x))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("The data contains NA as expression values. NAs are not allowed", " in the input.",</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">48</td>
                    <td class="coverage">25<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(x) != length(genes)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("The number of rows is not equal to the given gene set.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">50</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    what &lt;- match.arg(what)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # validate q values (Tsallis parameter must be &gt; 0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(q) || any(q &lt;= 0)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">55</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Argument 'q' must be numeric and greater than 0.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">57</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # keep a copy of transcript-level counts when available (non-null)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!exists("se_assay_mat")) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">59</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        se_assay_mat &lt;- x</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">60</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">61</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result &lt;- calculate_method(x, genes, norm, verbose = verbose, q = q, what = what)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">63</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">64</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Prepare assay and row/col data</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">65</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_assay &lt;- result[, -1, drop = FALSE]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_rowData &lt;- data.frame(genes = result[, 1], row.names = result[, 1])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">67</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(q) &gt; 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        col_split &lt;- do.call(rbind, strsplit(colnames(result)[-1], "_q="))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        col_ids &lt;- paste(col_split[, 1], "_q=", col_split[, 2], sep = "")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">71</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        row_ids &lt;- as.character(result[, 1])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">72</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_colData &lt;- data.frame(samples = as.character(col_split[, 1]), q = as.numeric(col_split[,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">73</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            2]), row.names = col_ids, stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">74</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        colnames(result_assay) &lt;- col_ids</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">75</td>
                    <td class="coverage">8<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(result_assay) &lt;- row_ids</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">77</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        col_ids &lt;- colnames(x)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        row_ids &lt;- as.character(result[, 1])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">79</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        result_colData &lt;- data.frame(samples = col_ids, row.names = col_ids)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        colnames(result_assay) &lt;- col_ids</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">13<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        rownames(result_assay) &lt;- row_ids</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">83</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_metadata &lt;- list(method = "tsallis", norm = norm, q = q, what = what)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">85</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">86</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # if we have preserved the original transcript-level matrix, build a</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">87</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # tx-&gt;gene mapping and make both available in metadata so downstream</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">88</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # plotting helpers can find transcript IDs and their parent genes</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    tx2gene_map &lt;- NULL</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (exists("se_assay_mat") &amp;&amp; !is.null(rownames(se_assay_mat)) &amp;&amp; length(genes) ==</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">91</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        nrow(se_assay_mat)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        tx2gene_map &lt;- data.frame(Transcript = rownames(se_assay_mat), Gen = as.character(genes),</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">95</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    assays_list &lt;- list()</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (what == "S") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">98</td>
                    <td class="coverage">20<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assays_list$diversity &lt;- result_assay</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (what == "D") {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">100</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        assays_list$hill &lt;- result_assay</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">101</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">102</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">103</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result &lt;- SummarizedExperiment::SummarizedExperiment(assays = assays_list, rowData = result_rowData,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        colData = result_colData, metadata = c(result_metadata, list(readcounts = if (exists("se_assay_mat")) se_assay_mat else NULL,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            tx2gene = tx2gene_map)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">107</td>
                    <td class="coverage">21<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(result)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">108</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/plot_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Internal plot helpers</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_format_label &lt;- function(lbl) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">4</td>
                    <td class="coverage">39<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(lbl)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(NULL)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    s &lt;- gsub("_", " ", lbl)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    s &lt;- gsub("\\s+", " ", s)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">9</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    s &lt;- trimws(s)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    s &lt;- tolower(s)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">11</td>
                    <td class="coverage">38<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nchar(s) == 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">12</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(s)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">14</td>
                    <td class="coverage">37<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nchar(s) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">15</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(toupper(s))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">16</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">17</td>
                    <td class="coverage">36<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    paste0(toupper(substr(s, 1, 1)), substr(s, 2, nchar(s)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">18</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">19</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_ma_plot_df &lt;- function(df, fold_col, mean_cols, x_label, y_label) {</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Detect x-axis values</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">22</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (length(mean_cols) &gt;= 2) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">23</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- rowMeans(df[, mean_cols[seq_len(2)], drop = FALSE], na.rm = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">24</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% paste0(mean_cols[1], " vs ", mean_cols[2])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">25</td>
                    <td class="coverage">12<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if (length(mean_cols) == 1) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">26</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- as.numeric(df[[mean_cols[1]]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% mean_cols[1]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">11<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else if ("mean" %in% colnames(df)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- as.numeric(df$mean)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">30</td>
                    <td class="coverage">9<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% "Mean"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">31</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        xvals &lt;- seq_len(nrow(df))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">33</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x_label &lt;- x_label %||% "Index"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">35</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    yvals &lt;- as.numeric(df[[fold_col]])</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">37</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_candidates &lt;- c("adjusted_p_values", "adj_p_value", "adj_p", "padj", "p.adjust")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_col &lt;- intersect(padj_candidates, colnames(df))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_col &lt;- if (length(padj_col))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        padj_col[1] else NULL</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">42</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj &lt;- if (!is.null(padj_col))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        as.numeric(df[[padj_col]]) else rep(1, length(yvals))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj[is.na(padj)] &lt;- 1</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">46</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">47</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sig_flag &lt;- ifelse(abs(yvals) &gt; 0 &amp; padj &lt; 0.05, "significant", "non-significant")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">48</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    plot_df &lt;- data.frame(genes = df$genes, x = xvals, y = yvals, padj = padj, significant = sig_flag,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stringsAsFactors = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">51</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">52</td>
                    <td class="coverage">15<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(plot_df = plot_df, x_label = x_label, y_label = y_label)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">53</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">54</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_prepare_volcano_df &lt;- function(diff_df, x_col = NULL, padj_col = "adjusted_p_values",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">56</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    label_thresh = 0.1, padj_thresh = 0.05, title = NULL) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">57</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- as.data.frame(diff_df)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">58</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    cn &lt;- colnames(df)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">59</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">60</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Auto-detect x-axis column if not specified</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">61</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(x_col)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">62</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        diff_cols &lt;- grep("_difference$", cn, value = TRUE, ignore.case = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">63</td>
                    <td class="coverage">5<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(diff_cols) &gt; 0) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">64</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            x_col &lt;- diff_cols[1]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">65</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">66</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            numeric_cols &lt;- vapply(df, is.numeric, logical(1))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">67</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            p_cols &lt;- grep("p_value|p.value", cn, ignore.case = TRUE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">68</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            numeric_cols[p_cols] &lt;- FALSE</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">69</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            if (any(numeric_cols)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">70</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">                x_col &lt;- cn[which(numeric_cols)[1]]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">71</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            } else {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">72</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">                stop("Could not find suitable column for x-axis. Specify 'x_col' explicitly.")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">73</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">74</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">75</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">76</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">77</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # Verify columns</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">78</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(x_col %in% cn))</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">79</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(sprintf("Column '%s' not found in diff_df", x_col))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">80</td>
                    <td class="coverage">10<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!(padj_col %in% cn))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">81</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop(sprintf("Column '%s' not found in diff_df", padj_col))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">82</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">83</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$xval &lt;- as.numeric(df[[x_col]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">84</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$padj &lt;- as.numeric(df[[padj_col]])</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">85</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$padj[is.na(df$padj)] &lt;- 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">86</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$padj[df$padj &lt;= 0] &lt;- .Machine$double.xmin</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">87</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">88</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df$significant &lt;- ifelse(abs(df$xval) &gt;= label_thresh &amp; df$padj &lt; padj_thresh,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">89</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "significant", "non-significant")</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">90</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    df &lt;- df[is.finite(df$xval) &amp; is.finite(df$padj), ]</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">91</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">92</td>
                    <td class="coverage">7<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(df) == 0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">93</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("No valid points to plot")</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">94</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">95</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    metric_label &lt;- if (grepl("median", x_col, ignore.case = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">96</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "Median" else if (grepl("mean", x_col, ignore.case = TRUE))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">97</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        "Mean" else "Value"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">98</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">99</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    title_use &lt;- title %||% "Volcano plot: fold-change vs significance"</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">100</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">101</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x_label_formatted &lt;- .tsenat_format_label(x_col)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">102</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    padj_label_formatted &lt;- .tsenat_format_label(padj_col)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">103</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">104</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    list(df = df, x_col = x_col, padj_col = padj_col, x_label_formatted = x_label_formatted,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">105</td>
                    <td class="coverage">6<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        padj_label_formatted = padj_label_formatted, title_use = title_use)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">106</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/calculate_method.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' Calculate Tsallis diversity values for transcripts grouped by gene</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param x Numeric matrix or data.frame of transcript-level expression</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' values (rows = transcripts, columns = samples).</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">4</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param genes Character vector with length equal to nrow(x) assigning each</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">5</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' transcript to a gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">6</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param norm Logical; if TRUE normalize Tsallis entropy values per gene.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">7</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param q Numeric scalar or vector of q values to evaluate.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">8</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param verbose Logical; show diagnostic messages when TRUE.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @param what Which quantity to return from `calculate_tsallis_entropy`:</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">10</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' 'S' (Tsallis entropy) or 'D' (Hill numbers) (default: 'S').</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">11</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @return A data.frame with genes in the first column and per-sample (and</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' per-q) Tsallis entropy values in subsequent columns.</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">#' @import stats</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">14</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">calculate_method &lt;- function(x, genes, norm = TRUE, verbose = FALSE, q = 2, what = c("S",</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    "D")) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">16</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    what &lt;- match.arg(what)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">17</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # validate q</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">18</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (!is.numeric(q) || any(q &lt;= 0)) {</pre>
                    </td>
                  </tr>
                  <tr class="missed">
                    <td class="num">19</td>
                    <td class="coverage">!</td>
                    <td class="col-sm-12">
                      <pre class="language-r">        stop("Argument 'q' must be numeric and greater than 0.", call. = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">20</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">21</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # cannot use aggregate because calculate_tsallis_entropy may return</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">22</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # multiple values when length(q) &gt; 1</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">23</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    gene_levels &lt;- unique(genes)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">24</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # ensure column names order matches the order used when constructing the</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">25</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # result matrix (samples vary outer, q varies inner). If sample names are</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">26</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # missing, synthesize deterministic names so column creation still works.</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">27</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    sample_names &lt;- colnames(x)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">28</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (is.null(sample_names)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">29</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        sample_names &lt;- paste0("Sample", seq_len(ncol(x)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">30</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">31</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    coln &lt;- as.vector(t(outer(sample_names, q, function(s, qq) paste0(s, "_q=", qq))))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">32</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rown &lt;- gene_levels</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">33</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">34</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    # compute requested quantity ('S' or 'D')</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">35</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    result_mat &lt;- t(vapply(gene_levels, function(gene) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">36</td>
                    <td class="coverage">91<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        .tsenat_tsallis_row(x = x, genes = genes, gene = gene, q = q, norm = norm,</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">37</td>
                    <td class="coverage">91<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            what = what)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">38</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }, FUN.VALUE = setNames(numeric(length(coln)), coln)))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">39</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(result_mat) &lt;- coln</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">40</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    rownames(result_mat) &lt;- rown</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">41</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    out_df &lt;- data.frame(Gene = rown, result_mat, check.names = FALSE)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">42</td>
                    <td class="coverage">27<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (all(rowSums(!is.na(result_mat)) == 0)) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">43</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        out_df &lt;- data.frame(Gene = character(0))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">44</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        for (nm in coln) out_df[[nm]] &lt;- numeric(0)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">45</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        x &lt;- out_df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">46</td>
                    <td class="coverage">3<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        return(x)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">47</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">48</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    x &lt;- out_df</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">49</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    y &lt;- x[apply(x[2:ncol(x)], 1, function(X) all(is.finite(X))), ]</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">50</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    if (nrow(x) - nrow(y) &gt; 0 &amp;&amp; verbose == TRUE) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">51</td>
                    <td class="coverage">1<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        message(sprintf("Note: %d genes excluded.", nrow(x) - nrow(y)))</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">52</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">53</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    colnames(y)[1] &lt;- "Gene"</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">54</td>
                    <td class="coverage">24<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    return(y)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">55</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div id="R/calculate_method_helpers.R" class="hidden">
              <table class="table-condensed">
                <tbody>
                  <tr class="never">
                    <td class="num">1</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"># Internal helpers for calculate_method</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">2</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r"></pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">3</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">.tsenat_tsallis_row &lt;- function(x, genes, gene, q, norm, what) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">4</td>
                    <td class="coverage">97<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    idx &lt;- which(genes == gene)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">5</td>
                    <td class="coverage">97<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    out &lt;- unlist(lapply(seq_len(ncol(x)), function(j) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">6</td>
                    <td class="coverage">343<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        v &lt;- calculate_tsallis_entropy(x[idx, j], q = q, norm = norm, what = what)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">7</td>
                    <td class="coverage">343<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        if (length(v) == length(q) &amp;&amp; all(is.finite(v) | is.na(v))) {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">8</td>
                    <td class="coverage">341<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            v</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">9</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        } else {</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">10</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            names_vec &lt;- paste0("q=", q)</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">11</td>
                    <td class="coverage">2<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">            setNames(rep(NA_real_, length(q)), names_vec)</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">12</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">        }</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">13</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    }))</pre>
                    </td>
                  </tr>
                  <tr class="covered">
                    <td class="num">14</td>
                    <td class="coverage">97<em>x</em></td>
                    <td class="col-sm-12">
                      <pre class="language-r">    out</pre>
                    </td>
                  </tr>
                  <tr class="never">
                    <td class="num">15</td>
                    <td class="coverage"></td>
                    <td class="col-sm-12">
                      <pre class="language-r">}</pre>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <script>$('div#files pre').each(function(i, block) {
    hljs.highlightBlock(block);
});</script>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
</body>
</html>
